---
title: "Visualisation of DESeq2 results with scatter plots"
description: "DEG analysis based on DESeq2 and GSEA"
principal investigator: "Joaquín de Navascués"
researcher: "Joaquín de Navascués"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: 'hide'
    theme: readable
    df_print: paged
---
```{r set-publication-theme, echo=FALSE, cache=FALSE}
ggplot2::theme_set(ggpubr::theme_pubr(base_size=10))
```

```{r setup, echo = FALSE, cache = FALSE}
knitr::opts_chunk$set(dev = 'png', 
                      fig.align = 'center', fig.height = 5, fig.width = 8.5, 
                      pdf.options(encoding = "ISOLatin9.enc"),
                      fig.path='integration/figures/', warning=FALSE, message=FALSE)
```

**Libraries needed:**
```{r load_libraries, warning=FALSE, echo=FALSE}
if (!require("librarian")) install.packages("librarian")
# data
librarian::shelf(readxl, dplyr, purrr, stringr, santoku, calibrate)
# graphics
librarian::shelf(cetcolor, ggplots2, ggpubr, ggtheme, ggtext, ggrepel)
# convenience
librarian::shelf(here)
setwd(here::here()) # to distinguish from dplyr::here()
```

**Path to definitive images (outside repo):**
```{r define_dir2figs}
figdir <- paste0(c(head(str_split(getwd(),'/')[[1]],-1),
                   paste0(tail(str_split(getwd(),'/')[[1]],1), '_figures')),
                 collapse='/')
dir.create(figdir, showWarnings = FALSE)
```


# RNAseq results visualisation with scatter plots


## 1 Getting ready


This gets us the DGE data from `DESeq2`, identified by FlyBase/Ensembl ID and gene symbol:
```{r load_DEG_data}
# experimental design and labels
targets <- readRDS('output/targets.RDS')
# DEG data
DaKD_deg <- readRDS('output/Control_vs_DaKD.RDS')
DaOE_deg <- readRDS('output/Control_vs_DaOE.RDS')
DaDaOE_deg <- readRDS('output/Control_vs_DaDaOE.RDS')
ScOE_deg <- readRDS('output/Control_vs_ScOE.RDS')
head(ScOE_deg, 1)
```

As discussed previously, we do not really need TPM values here, as the `baseMean` from the `DESeq2` normalisation does the work just as well and is more 'internally consistent'.

Now make a data frame with the status of each gene (up/down/non-regulated) for each condition:
```{r extract_regulated_sets2, warning=FALSE}
# first define the function
extract_regulated_sets2 <- function(list_of_degs, names_degs, fc_thresh=1.5, cols) {
  # fc_thresh must be a positive number
  # list_of_degs is a list of dataframes
  # names_degs is a list of strings
  # they must have the same length
  reg_lvl <- paste0(' reg@log~2~FC≥', fc_thresh)
  breaks <- brk_manual(c(-fc_thresh, fc_thresh), left_vec = c(FALSE, TRUE))
  regulated_sets <- NULL
  for (l in 1:length(list_of_degs)) {
    degs_na <- dplyr::select(list_of_degs[[l]], all_of(cols))
    # better not pass NAs to kiru
    degs <- na.omit(degs_na)
    # filter by log2FC threshold, create `reg`(ulated) col with up/down/non
    degs$reg <- 
      kiru(
        degs$log2FoldChange,
        breaks = breaks,
        extend = TRUE,
        labels=c("down", "non", "up")
      )
    # filter by p-val
    degs$reg <- as.character(degs$reg) # remove factor
    x <- 1:nrow(degs)
    degs$reg <- ifelse(degs[x,'padj']<0.05, degs[x,'reg'], 'non')
    # recover NAs as non-regulataed
    added_nas <- degs_na[!(rownames(degs_na) %in% rownames(na.omit(degs_na))),]
    added_nas$reg <- 'non'
    degs <- rbind(degs, added_nas)
    # store reg status
    degs[ , names_degs[[l]] ] <- degs$reg
    regulated_sets[[l]] <- dplyr::select(degs, c(ensemblGeneID, names_degs[[l]]) )
  }
  regulated_sets <- purrr::reduce(regulated_sets, full_join, by='ensemblGeneID')
  rownames(regulated_sets) <- regulated_sets$ensemblGeneID
  regulated_sets <- dplyr::select(regulated_sets, -ensemblGeneID)
  return(regulated_sets)
}

# now use the function 
list_of_degs <- list('kd'=DaKD_deg, 'da'=DaOE_deg,
                     'dada'=DaDaOE_deg, 'sc'=ScOE_deg)
names_degs <- list('DaKD','DaOE','DaDaOE','ScOE')
cols.oi <- c('log2FoldChange', 'padj', 'baseMean',
             'gene_symbol', 'ensemblGeneID')
fc_thresh <- 1.5
regs <- extract_regulated_sets2(list_of_degs, names_degs, fc_thresh=fc_thresh, cols=cols.oi)

# attach the results to the DEG data frame for each condition
DaKD_deg$reg <- regs$DaKD
DaOE_deg$reg <- regs$DaOE
DaDaOE_deg$reg <- regs$DaDaOE
ScOE_deg$reg <- regs$ScOE
```

Load `genes_of_interest`:
```{r load_GOI}
genes_of_interest <- read_excel('resources/genes_of_interest.xlsx', sheet='core')
```


## 2 MA plots using `ggpubr`

Customise the looks so it matches the prior version of the paper:
```{r MA_plot_customisation}
ggmaplot2 <- function(deg, goi_list, fc_thresh = 1.5,
                      md_label='*genotype^—^*', repulsion) {
  # deg is the output of DESeq2 with an added column of gene symbols
  # gene.symbols is that column
  
  # standard MA plot with `ggpubr`, using the CBD1 palette from `cetcolor` 
  ma <- ggmaplot(
    deg,
    fdr = 0.05,
    fc = fc_thresh,
    genenames = deg$gene_symbols,
    size = 2,
    alpha = 0.5,
    seed = NA,
    font.label = c(16, "bold", "black"),
    label.rectangle = FALSE,
    palette = c(cet_pal(n = 3, name = "cbd1", alpha = 1)[3], # "#A89008"
                cet_pal(n = 3, name = "cbd1", alpha = 1)[1], # "#3A90FE"
                "#AAAAAA"),
    top = 0,
    main = NULL,
    xlab = "log~2~(mean expression)",
    ylab = "log~2~(fold change)",
    ggtheme = theme_linedraw(),
    legend = 'top'
  )
  
  annotations <- data.frame(
        xpos = floor(min(ggplot_build(ma)$layout$panel_params[[1]]$x.range)),
        ypos = ceiling(max(abs(ggplot_build(ma)$layout$panel_params[[1]]$y.range))),
        annotateText = md_label,
        hjustvar = 0,
        vjustvar = 1)
  
  return(ma +
    # mark special genes if upreg
    geom_text_repel(data = deg %>% filter(gene_symbol %in% goi_list$gene_symbol &
                                               log2FoldChange > 0),
                    aes(x = log2(baseMean),
                        y = log2FoldChange,
                        label = gene_symbol,
                        segment.square  = FALSE,
                        segment.inflect = TRUE),
                    color         = '#6E5C03', # ==hue, <ligthness than than "#A89008"
                    fontface = 'bold',
                    segment.alpha = 0.8,
                    segment.linetype = 3,
                    hjust = 0,
                    # positioning
                    box.padding   = repulsion$box.padding,
                    point.padding = repulsion$point.padding,
                    nudge_x       = repulsion$nudge_x.up,
                    nudge_y       = repulsion$nudge_y.up,
                    force         = repulsion$force,
                    force_pull    = repulsion$force_pull,
                    max.overlaps  = Inf,
                    xlim          = repulsion$xlims.up,    # NA repels from edges
                    ylim          = repulsion$ylims.up,
                    seed = 42) +
    # mark special genes if downreg
    geom_text_repel(data = deg %>% filter(gene_symbol %in% goi_list$gene_symbol &
                                               log2FoldChange < 0),
                    aes(x = log2(baseMean),
                        y = log2FoldChange,
                        label = gene_symbol,
                        segment.square  = FALSE,
                        segment.inflect = TRUE),
                     color         = '#0565AF', # ==hue, <ligthness than "#3A90FE"
                     fontface = 'bold',
                     segment.alpha = 0.8,
                     segment.linetype = 3,
                     hjust = 0,
                     # positioning
                     box.padding   = repulsion$box.padding,
                     point.padding = repulsion$point.padding,
                     nudge_x       = repulsion$nudge_x.dn,
                     nudge_y       = repulsion$nudge_y.dn,
                     force         = repulsion$force,
                     force_pull    = repulsion$force_pull,
                     max.overlaps  = Inf,
                     xlim          = repulsion$xlims.dn,             # NA repels from edges
                     ylim          = repulsion$ylims.dn,
                     seed = 42) +
    # genotype label
    geom_richtext(
      data = annotations,
      aes(x = xpos,y = ypos,
          hjust = hjustvar, vjust = vjustvar,
          label = annotateText,
          fontface = 'bold',
          size = 9),
      fill = NA,
      label.color = NA,
      label.padding = grid::unit(rep(0, 4), "pt"),
      label.margin = grid::unit(rep(0, 4), "pt")
      ) +
    # legend
    guides(size = 'none') +
    # apply markdown formatting, etc
    theme(# markdown
          axis.title.x = element_markdown(size = 14, face = 'bold'),
          axis.title.y = element_markdown(size = 14, face = 'bold'),
          axis.text = element_text(size = 10),
          # grid and panel
          panel.border = element_rect(color = "black", fill = NA, linewidth = 1),
          panel.grid = element_blank(),
          # legend
          legend.margin = margin(b = -10),
          legend.key.size = unit(1.2, 'cm'),
          legend.text = element_text(size = 12, face = 'bold')
          )
    )
}
```

Apply this to the *daughterless* knockdown:
```{r ggmaplot2_da_knockdown}
# define genes of interest
# GoI <- genes_of_interest %>%
#   filter(`FlyCellAtlas marker`=='T') %
#   filter((`fate promoted` == 'EE') | 
#          (`fate promoted` == 'EC')) %>%
#   dplyr::select(gene_symbol)
  
# genes of interest dysregulated in this condition
goi_kd <- DaKD_deg %>% dplyr::filter(
  gene_symbol %in% genes_of_interest$gene_symbol &
  reg!='non' &
  abs(log2FoldChange)>1.5 &
  padj<0.05
  )

repulsion <- list(box.padding = 0.5,
                  point.padding = 0.8,
                  nudge_x = 3,
                  nudge_y = 2,
                  force = 2,
                  force_pull = 0.2,
                  xlims.up = c(18, NA),
                  ylims.up = c(10, NA),
                  xlims.dn = c(10, NA),
                  ylims.dn = c(-10, NA))

ggmaplot2 (deg = DaKD_deg,
           fc_thresh = fc_thresh,
           goi_list = goi_kd, 
           md_label = '*esg > da^RNAi^*',
           repulsion = repulsion)
```

```{r, echo=FALSE, include=FALSE}
ggsave('MA_daKD.pdf', plot = last_plot(), device = 'pdf',
       path = figdir, dpi = 300)
ggsave('MA_daKD.svg', plot = last_plot(), device = 'svg',
       path = figdir, dpi = 300)
ggsave('MA_daKD.png', plot = last_plot(), device = png,
       path = figdir, dpi = 300)
```

To the *daughterless* overexpression:
```{r ggmaplot2_da_overexpression}
# genes of interest dysregulated in this condition
goi_da <- DaOE_deg %>% dplyr::filter(
  gene_symbol %in% genes_of_interest$gene_symbol &
  reg!='non' &
  abs(log2FoldChange)>2 &
  padj<0.05
  )

repulsion <- list(box.padding = 0.8,
                  point.padding = 0.5,
                  force = 2,
                  force_pull = 0.05,
                  nudge_x.up = 3,
                  nudge_y.up = 2,
                  xlims.up = c(5, NA),
                  ylims.up = c(7, NA),
                  nudge_x.dn = 3,
                  nudge_y.dn = -2,
                  xlims.dn = c(12, NA),
                  ylims.dn = c(NA, -3))

ggmaplot2(deg = DaOE_deg,
          fc_thresh = fc_thresh,
          goi_list = goi_da, 
          md_label = '*esg > da*',
          repulsion = repulsion)
```

```{r, echo=FALSE, include=FALSE}
suppressMessages(ggsave('MA_daOE.pdf', plot = last_plot(), device = 'pdf',
                        path = figdir, dpi = 300))
suppressMessages(ggsave('MA_daOE.svg', plot = last_plot(), device = 'svg',
                        path = figdir, dpi = 300))
suppressMessages(ggsave('MA_daOE.png', plot = last_plot(), device = png,
                        path = figdir, dpi = 300))
```

To the Daughterless homodimer overexpression:
```{r ggmaplot2_da:da_overexpression}
# genes of interest dysregulated in this condition
goi_dada <- DaDaOE_deg %>% dplyr::filter(
  gene_symbol %in% genes_of_interest$gene_symbol &
  reg!='non' &
  abs(log2FoldChange)>2 &
  padj<0.05
  )

repulsion <- list(box.padding = 0.8,
                  point.padding = 0.5,
                  force = 2,
                  force_pull = 0.1,
                  nudge_x.up = 3,
                  nudge_y.up = 2,
                  xlims.up = c(5, NA),
                  ylims.up = c(7, NA),
                  nudge_x.dn = 5,
                  nudge_y.dn = -3,
                  xlims.dn = c(10, NA),
                  ylims.dn = c(NA, -1))

ggmaplot2(deg = DaDaOE_deg,
          fc_thresh = fc_thresh,
          goi_list = goi_dada, 
          md_label = '*esg > da:da*',
          repulsion = repulsion)
```

```{r, echo=FALSE, include=FALSE}
suppressMessages(ggsave('MA_dadaOE.pdf', plot = last_plot(), device = 'pdf',
                        path = figdir, dpi = 300))
suppressMessages(ggsave('MA_dadaOE.svg', plot = last_plot(), device = 'svg',
                        path = figdir, dpi = 300))
suppressMessages(ggsave('MA_dadaOE.png', plot = last_plot(), device = png,
                        path = figdir, dpi = 300))
```

And, finally, to the *scute* overexpression:
```{r ggmaplot2_sc_overexpression}
# genes of interest dysregulated in this condition
goi_sc <- ScOE_deg %>% dplyr::filter(
  gene_symbol %in% genes_of_interest$gene_symbol &
  reg!='non' &
  abs(log2FoldChange)>2 &
  padj<0.05
  )

repulsion <- list(box.padding = 0.8,
                  point.padding = 0.5,
                  force = 2,
                  force_pull = 0.1,
                  nudge_x.up = 3,
                  nudge_y.up = 2,
                  xlims.up = c(5, NA),
                  ylims.up = c(7, NA),
                  nudge_x.dn = 5,
                  nudge_y.dn = -3,
                  xlims.dn = c(NA, NA),
                  ylims.dn = c(NA, -5))

ggmaplot2(deg = ScOE_deg,
          fc_thresh = fc_thresh,
          goi_list = goi_sc, 
          md_label = '*esg > scute*',
          repulsion = repulsion)
```

```{r, echo=FALSE, include=FALSE}
suppressMessages(ggsave('MA_ScOE.pdf', plot = last_plot(), device = 'pdf',
                        path = figdir, dpi = 300))
suppressMessages(ggsave('MA_ScOE.svg', plot = last_plot(), device = 'svg',
                        path = figdir, dpi = 300))
suppressMessages(ggsave('MA_ScOE.png', plot = last_plot(), device = png,
                        path = figdir, dpi = 300))
```
