---
title: "Analysis of DESeq2 results with gene set enrichment"
description: "DEG analysis based on DESeq2 and GSEA"
principal investigator: "Joaquín de Navascués"
researcher: "Joaquín de Navascués"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: 'hide'
    theme: readable
    df_print: paged
---

# Analysis of DESeq2 results with gene set enrichment

## 0 Boilerplate

```{r set-publication-theme, echo=FALSE, cache=FALSE}
ggplot2::theme_set(ggpubr::theme_pubr(base_size=10))
```

```{r setup, echo = FALSE, cache = FALSE}
knitr::opts_chunk$set(dev = 'png', 
                      fig.align = 'center', fig.height = 7, fig.width = 8.5, 
                      pdf.options(encoding = "ISOLatin9.enc"),
                      fig.path='integration/figures/', warning=FALSE, message=FALSE)
```

**Libraries needed:**
```{r load_libraries, warning=FALSE, echo=FALSE}
if (!require("librarian")) install.packages("librarian")
# data
librarian::shelf(
  # data manip.
  dplyr, tidyr, stringr, org.Dm.eg.db, DOSE, fgsea, clusterProfiler, purrr, biomaRt, gage,
  # plotting
  pathview,  cetcolor, enrichplot, genekitr, ggh4x, ggtheme, ggtext, ggrepel, ggnewscale, patchwork, pathview,
# convenience
  here)
```

**Set working directory where the script is**
```{r setwd}
if (Sys.getenv("USER")=="JQ") {
  setwd("/Users/JQ/Documents/_CODE REPOS/GitHub/Da_RNAseq")
} else if (Sys.getenv("RSTUDIO")==1) {
  setwd( dirname(rstudioapi::getSourceEditorContext(id = NULL)$path) ) # gets what is in the editor
} else {
  setwd(here::here())
  d <- str_split(getwd(),'/')[[1]][length(str_split(getwd(),'/')[[1]])]
  if (d != 'Da_RNAseq') { stop(
    paste0("Could not set working directory automatically to where this",
           " script resides.\nPlease do `setwd()` manually"))
    }
}
getwd()
```

**Path to definitive images (outside repo):**
```{r define_dir2figs}
figdir <- paste0(c(head(str_split(getwd(),'/')[[1]],-1),
                   paste0(tail(str_split(getwd(),'/')[[1]],1), '_figures')),
                 collapse='/')
dir.create(figdir, showWarnings = FALSE)
```




## 1 Getting ready


This gets us the DGE data from `DESeq2`, identified by FlyBase/Ensembl ID and gene symbol:
```{r load_DEG_data}
# experimental design and labels
targets <- readRDS('output/targets.RDS')
# DEG data
DaKD_deg <- readRDS('output/Control_vs_DaKD.RDS')
DaOE_deg <- readRDS('output/Control_vs_DaOE.RDS')
DaDaOE_deg <- readRDS('output/Control_vs_DaDaOE.RDS')
ScOE_deg <- readRDS('output/Control_vs_ScOE.RDS')
head(
  ScOE_deg %>% dplyr::select(gene_symbol, ensemblGeneID, baseMean, log2FoldChange, padj),
  1
)
```

To get a differentially expressed gene set for Over-Representation Analysis:
```{r define_degset}
make_degset <- function(deg, up, fc_thresh) {
  converter <- ifelse(up,1,-1)
  degset <- deg %>%
    filter(padj<0.05) %>%
    filter(abs(log2FoldChange) > fc_thresh) %>%
    filter(log2FoldChange*converter > 0)
 return(degset)
}
```

To get a ranked gene list for Gene Set Enrichment Analysis:
```{r define_degrank}
make_degrank <- function(deg, mode='log2fc', key='gene_symbol') {
  if (!(key %in% c('gene_symbol', 'ensemblGeneID'))) {
    stop("The `key` parameter needs to be one of 'gene_symbol', 'ensemblGeneID'")
  }
  # mode: whether to use padj or log2FC
  if (mode=='log2fc') {
    deg_sorted <- deg %>% arrange(desc(log2FoldChange))
    degrank <- deg_sorted$log2FoldChange
    names(degrank) <- deg_sorted[,key]
  } else if (mode=='padj') {
    deg$p.rank <- sign(deg$log2FoldChange) * -log10(deg$padj)
    deg_sorted <- deg %>% arrange(desc(p.rank))
    degrank <- deg_sorted$p.rank
    names(degrank) <- deg_sorted[,key]
  } else {
    stop('the `mode` parameter can only be either `log2fc` or `padj`')
  }
 return(degrank)
}
```


## 2 Over-representation of Gene Ontology terms 


For *daughterless* knockdown:
```{r ORA_da_knockdown}
# using `clusterProfiler` and `enrichplot`
# select reg set
daKD_up <- make_degset(DaKD_deg, up=TRUE, fc_thresh=1.5)
# ORA
daKD_up_eGO <- enrichGO(gene         = daKD_up$gene_symbol,
                        OrgDb         = org.Dm.eg.db,
                        keyType       = 'SYMBOL',
                        pAdjustMethod = "BH",
                        pvalueCutoff  = 0.05,
                        qvalueCutoff  = 0.05)
# with down genes
daKD_dn <- make_degset(DaKD_deg, up=FALSE, fc_thresh=1.5)
daKD_dn_eGO <- enrichGO(gene         = daKD_dn$gene_symbol,
                        OrgDb         = org.Dm.eg.db,
                        keyType       = 'SYMBOL',
                        pAdjustMethod = "BH",
                        pvalueCutoff  = 0.05,
                        qvalueCutoff  = 0.05)
```

```{r ORA_da_knockdown_plots}
# plot
up <- dotplot(daKD_up_eGO, showCategory=30, label_format=50) +
  ggtitle("Over-represented GO terms — genes **up** in *esg > da^RNAi^*") +
  scale_colour_gradient(low = cet_pal(3, name='cbd1')[2],
                        high = cet_pal(3, name='cbd1')[3]) +
  theme(plot.title = element_markdown(size=rel(1.2)),
        panel.grid = element_blank(),
        axis.text.y = element_text(size=8, lineheight=0.5))

down <- dotplot(daKD_dn_eGO, showCategory=30, label_format=40) +
  ggtitle("Over-represented GO terms — genes **down** in *esg > da^RNAi^*") +
  scale_colour_gradient(low = cet_pal(3, name='cbd1')[2],
                        high = cet_pal(3, name='cbd1')[1]) +
  theme(plot.title = element_markdown(),
        panel.grid = element_blank(),
        axis.text.y = element_text(size=9))

suppressWarnings(print(up))
suppressWarnings(print(down))
```

```{r dotplots_using_`genekitr`, include=FALSE}
# plotEnrich(importCP(daKD_up_eGO, type='go'),
#            plot_type = 'dot',
#            up_color = cet_pal(3,name='cbd1')[3],
#            down_color = cet_pal(3,name='cbd1')[2])
# 
# plotEnrich(importCP(daKD_dn_eGO, type='go'),
#            plot_type = 'dot',
#            up_color = cet_pal(3,name='cbd1')[1],
#            down_color = cet_pal(3,name='cbd1')[2])
# plotEnrichAdv(importCP(daKD_up_eGO, type='go'), importCP(daKD_dn_eGO, type='go'),
#               plot_type = 'one',
#               term_metric = "FoldEnrich",
#               stats_metric = "p.adjust",
#               color =c(cet_pal(2,name='cbd1')[1], cet_pal(2,name='cbd1')[2]))
```

For *daughterless* overexpression:
```{r ORA_da_overexpression}
# using `clusterProfiler` and `enrichplot`
# select reg set
daOE_up <- make_degset(DaOE_deg, up=TRUE, fc_thresh=1.5)
# ORA
daOE_up_eGO <- enrichGO(gene         = daOE_up$gene_symbol,
                        OrgDb         = org.Dm.eg.db,
                        keyType       = 'SYMBOL',
                        pAdjustMethod = "BH",
                        pvalueCutoff  = 0.05,
                        qvalueCutoff  = 0.05)
# with down genes
daOE_dn <- make_degset(DaOE_deg, up=FALSE, fc_thresh=1.5)
daOE_dn_eGO <- enrichGO(gene         = daOE_dn$gene_symbol,
                        OrgDb         = org.Dm.eg.db,
                        keyType       = 'SYMBOL',
                        pAdjustMethod = "BH",
                        pvalueCutoff  = 0.05,
                        qvalueCutoff  = 0.05)
```
```{r ORA_da_overexpression_plots}
# plot
up <- dotplot(daOE_up_eGO, showCategory=30, label_format=60) +
  ggtitle("Over-represented GO terms — genes **up** in *esg > da*") +
  scale_colour_gradient(low = cet_pal(3, name='cbd1')[2],
                        high = cet_pal(3, name='cbd1')[3]) +
  theme(plot.title = element_markdown(size=rel(1.15)),
        panel.grid = element_blank(),
        axis.text.y = element_text(size=9))

down <- dotplot(daOE_dn_eGO, showCategory=30, label_format=60) +
  ggtitle("Over-represented GO terms — genes **down** in *esg > da*") +
  scale_colour_gradient(low = cet_pal(3, name='cbd1')[2],
                        high = cet_pal(3, name='cbd1')[1]) +
  theme(plot.title = element_markdown(size=rel(1.15)),
        panel.grid = element_blank(),
        axis.text.y = element_text(size=8, lineheight=0.5))

suppressWarnings(print(up))
suppressWarnings(print(down))
```

For *da:da* overexpression:
```{r ORA_dada_overexpression}
# using `clusterProfiler` and `enrichplot`
# select reg set
dadaOE_up <- make_degset(DaDaOE_deg, up=TRUE, fc_thresh=1.5)
# ORA
dadaOE_up_eGO <- enrichGO(gene         = dadaOE_up$gene_symbol,
                        OrgDb         = org.Dm.eg.db,
                        keyType       = 'SYMBOL',
                        pAdjustMethod = "BH",
                        pvalueCutoff  = 0.05,
                        qvalueCutoff  = 0.05)
# with down genes
dadaOE_dn <- make_degset(DaDaOE_deg, up = FALSE, fc_thresh = 1.5)
dadaOE_dn_eGO <- enrichGO(gene        = dadaOE_dn$gene_symbol,
                        OrgDb         = org.Dm.eg.db,
                        keyType       = 'SYMBOL',
                        pAdjustMethod = "BH",
                        pvalueCutoff  = 0.05,
                        qvalueCutoff  = 0.05)
```
```{r ORA_dada_overexpression_plots}
# plot
up <- dotplot(dadaOE_up_eGO, showCategory=30, label_format=40) +
  ggtitle("Over-represented GO terms — genes **up** in *esg > da:da*") +
  scale_colour_gradient(low = cet_pal(3, name='cbd1')[2],
                        high = cet_pal(3, name='cbd1')[3]) +
  force_panelsizes(rows = unit(3, "in"), cols = unit(2, "in")) +
  theme(plot.title = element_markdown(),
        plot.title.position = 'plot',
        panel.grid = element_blank(),
        axis.text.y = element_text(size=9))

down <- dotplot(dadaOE_dn_eGO, showCategory=30, label_format=40) +
  ggtitle("Over-represented GO terms — genes **down** in *esg > da:da*") +
  scale_colour_gradient(low = cet_pal(3, name='cbd1')[2],
                        high = cet_pal(3, name='cbd1')[1]) +
  force_panelsizes(rows = unit(3, "in"), cols = unit(2, "in")) +
  theme(plot.title = element_markdown(),
        plot.title.position = 'plot',
        panel.grid = element_blank(),
        axis.text.y = element_text(size=9))

tryCatch({ print(up) },
  error = function (e) {
    message('Error occurred when trying to plot dotplot')
    print(e)})

tryCatch({ print(down) },
  error = function (e) {
    message('Error occurred when trying to plot dotplot')
    print(e)})
```

For *scute* overexpression:
```{r ORA_sc_overexpression}
# using `clusterProfiler` and `enrichplot`
# select reg set
scOE_up <- make_degset(ScOE_deg, up=TRUE, fc_thresh=1.5)
# ORA
scOE_up_eGO <- enrichGO(gene         = scOE_up$gene_symbol,
                        OrgDb         = org.Dm.eg.db,
                        keyType       = 'SYMBOL',
                        pAdjustMethod = "BH",
                        pvalueCutoff  = 0.05,
                        qvalueCutoff  = 0.05)
# with down genes
scOE_dn <- make_degset(ScOE_deg, up=FALSE, fc_thresh=1.5)
scOE_dn_eGO <- enrichGO(gene         = scOE_dn$gene_symbol,
                        OrgDb         = org.Dm.eg.db,
                        keyType       = 'SYMBOL',
                        pAdjustMethod = "BH",
                        pvalueCutoff  = 0.05,
                        qvalueCutoff  = 0.05)
```
```{r ORA_sc_overexpression_plots}
# plot
up <- dotplot(scOE_up_eGO, showCategory=30, label_format=60) +
  ggtitle("Over-represented GO terms — genes **up** in *esg > scute*") +
  scale_colour_gradient(low = cet_pal(3, name='cbd1')[2],
                        high = cet_pal(3, name='cbd1')[3]) +
  theme(plot.title = element_markdown(),
        panel.grid = element_blank(),
        axis.text.y = element_text(size=9))

down <- dotplot(scOE_dn_eGO, showCategory=30, label_format=60) +
  ggtitle("Over-represented GO terms — genes **down** in *esg > scute*") +
  scale_colour_gradient(low = cet_pal(3, name='cbd1')[2],
                        high = cet_pal(3, name='cbd1')[1]) +
  theme(plot.title = element_markdown(),
        panel.grid = element_blank(),
        axis.text.y = element_text(size=8, lineheight=0.5))

suppressWarnings(print(up))
suppressWarnings(print(down))
```

None of these plots are particularly informative, so I am going to move on to GSEA and custom gene sets.


## 3 Gene Set Enrichment Analysis with Gene Ontology terms


For *daughterless* knockdown:
```{r GSEA_da_knockdown}
# using `clusterProfiler` and `enrichplot`
# select reg set
daKD_rank <- make_degrank(DaKD_deg, mode='log2fc')
# GSEA
daKD_GSEgo <- gseGO(geneList=daKD_rank,
                    keyType = 'SYMBOL',
                    OrgDb=org.Dm.eg.db,
                    minGSSize    = 15,
                    maxGSSize    = 1000,
                    pvalueCutoff = 0.05,
                    eps = 0,
                    verbose=F)
```

```{r GSEA_da_knockdown_plots}
# get the logP values instead of padj 
daKD_GSEgo_log <- daKD_GSEgo
daKD_GSEgo_log@result$p.adjust <- -log10(daKD_GSEgo_log@result$p.adjust) # use @slot and $column to assign
# to subset the data to the highest NES values
#daKD_GSEgo_log@result <- subset(daKD_GSEgo_log@result, abs(NES)>1.8)
# to subset the data to the lowest p-values:
number_of_terms <- 30
cutoff <- rev(rev(sort(daKD_GSEgo_log$p.adjust))[1:number_of_terms])[1]
daKD_GSEgo_log@result <- subset(daKD_GSEgo_log@result, p.adjust>=cutoff)
# to control colour of GO terms according to enrichment or depletion
d <- ifelse(sort(daKD_GSEgo_log$NES) > 0, cet_pal(2, name='cbd1')[2], cet_pal(2, name='cbd1')[1])
# plot
gsep <- dotplot(daKD_GSEgo_log,
                x='NES', 
                showCategory=number_of_terms,
                label_format=60,
                ) +
  ggtitle("Enriched/depleted GO terms in *esg > da^RNAi^*") +
  scale_colour_gradient(low = cet_pal(3, name='d2')[2],
                        high = cet_pal(3, name='d2')[3]) +
  labs(color = '-log~10~(*p*)') +
  theme(plot.title = element_markdown(),
        plot.title.position = 'panel',
        legend.title = element_markdown(),
        panel.grid = element_blank(),
        axis.text.y = element_text(size=9, colour=d))

suppressWarnings(print(gsep))
```

For *daughterless* overexpression:
```{r GSEA_da_overexpression}
# using `clusterProfiler` and `enrichplot`
# select reg set
daOE_rank <- make_degrank(DaOE_deg, mode='log2fc')
# GSEA
daOE_GSEgo <- gseGO(geneList=daOE_rank,
                    keyType = 'SYMBOL',
                    OrgDb=org.Dm.eg.db,
                    minGSSize    = 15,
                    maxGSSize    = 1000,
                    pvalueCutoff = 0.05,
                    eps = 0,
                    verbose=F)
```

```{r GSEA_da_overexpression_plots}
# get the logP values instead of padj 
daOE_GSEgo_log <- daOE_GSEgo
daOE_GSEgo_log@result$p.adjust <- -log10(daOE_GSEgo_log@result$p.adjust) # use @slot and $column to assign
# to subset the data to the highest NES values
#daOE_GSEgo_log@result <- subset(daOE_GSEgo_log@result, abs(NES)>1.8)
# to subset the data to the lowest p-values:
number_of_terms <- 30
cutoff <- rev(rev(sort(daOE_GSEgo_log$p.adjust))[1:number_of_terms])[1]
daOE_GSEgo_log@result <- subset(daOE_GSEgo_log@result, p.adjust>=cutoff)
# to control colour of GO terms according to enrichment or depletion
d <- ifelse(sort(daOE_GSEgo_log$NES) > 0, cet_pal(2, name='cbd1')[2], cet_pal(2, name='cbd1')[1])
# plot
gsep <- dotplot(daOE_GSEgo_log,
                x='NES', 
                showCategory=number_of_terms,
                label_format=60,
                ) +
  ggtitle("Enriched/depleted GO terms in *esg > da*") +
  scale_colour_gradient(low = cet_pal(3, name='d2')[2],
                        high = cet_pal(3, name='d2')[3]) +
  labs(color = '-log~10~(*p*)') +
  theme(plot.title = element_markdown(),
        plot.title.position = 'panel',
        legend.title = element_markdown(),
        panel.grid = element_blank(),
        axis.text.y = element_text(size=9, colour=d))

suppressWarnings(print(gsep))
```

For *da:da* overexpression:
```{r GSEA_dada_overexpression}
# using `clusterProfiler` and `enrichplot`
# select reg set
dadaOE_rank <- make_degrank(DaDaOE_deg, mode='log2fc')
# GSEA
dadaOE_GSEgo <- gseGO(geneList=dadaOE_rank,
                    keyType = 'SYMBOL',
                    OrgDb=org.Dm.eg.db,
                    minGSSize    = 15,
                    maxGSSize    = 1000,
                    pvalueCutoff = 0.05,
                    eps = 0,
                    verbose=F)
```

```{r GSEA_da_overexpression_plots}
# get the logP values instead of padj 
dadaOE_GSEgo_log <- dadaOE_GSEgo
dadaOE_GSEgo_log@result$p.adjust <- -log10(dadaOE_GSEgo_log@result$p.adjust) # use @slot and $column to assign
# to subset the data to the highest NES values
#dadaOE_GSEgo_log@result <- subset(dadaOE_GSEgo_log@result, abs(NES)>1.8)
# to subset the data to the lowest p-values:
number_of_terms <- 30
cutoff <- rev(rev(sort(dadaOE_GSEgo_log$p.adjust))[1:number_of_terms])[1]
dadaOE_GSEgo_log@result <- subset(dadaOE_GSEgo_log@result, p.adjust>=cutoff)
# to control colour of GO terms according to enrichment or depletion
d <- ifelse(sort(dadaOE_GSEgo_log$NES) > 0,
            cet_pal(2, name='cbd1')[2], cet_pal(2, name='cbd1')[1])
# plot
gsep <- dotplot(dadaOE_GSEgo_log,
                x='NES', 
                showCategory=number_of_terms,
                label_format=50,
                ) +
  ggtitle("Enriched/depleted GO terms in *esg > da:da*") +
  scale_colour_gradient(low = cet_pal(3, name='d2')[2],
                        high = cet_pal(3, name='d2')[3]) +
  labs(color = '-log~10~(*p*)') +
  #coord_fixed(ratio=0.8) +
  force_panelsizes(rows = unit(3.5,'in'), cols = unit(4,'in')) +
  theme(plot.title = element_markdown(),
        plot.title.position = 'panel',
        legend.title = element_markdown(),
        panel.grid = element_blank(),
        axis.text.y = element_text(size=7, lineheight=0.7, colour=d)
        )
#gsep$coordinates$ratio <- 0.01
suppressWarnings(print(gsep))
```

For *scute* overexpression:
```{r GSEA_dada_overexpression}
# using `clusterProfiler` and `enrichplot`
# select reg set
scOE_rank <- make_degrank(ScOE_deg, mode='log2fc')
# GSEA
scOE_GSEgo <- gseGO(geneList=scOE_rank,
                    keyType = 'SYMBOL',
                    OrgDb=org.Dm.eg.db,
                    minGSSize    = 15,
                    maxGSSize    = 1000,
                    pvalueCutoff = 0.05,
                    nPermSimple = 1000000,
                    eps = 0,
                    verbose=F)
```

```{r GSEA_da_overexpression_plots}
# get the logP values instead of padj 
scOE_GSEgo_log <- scOE_GSEgo
scOE_GSEgo_log@result$p.adjust <- -log10(scOE_GSEgo_log@result$p.adjust) # use @slot and $column to assign
# to subset the data to the highest NES values
#scOE_GSEgo_log@result <- subset(scOE_GSEgo_log@result, abs(NES)>1.8)
# to subset the data to the lowest p-values:
number_of_terms <- 30
cutoff <- rev(rev(sort(scOE_GSEgo_log$p.adjust))[1:number_of_terms])[1]
scOE_GSEgo_log@result <- subset(scOE_GSEgo_log@result, p.adjust>=cutoff)
# to control colour of GO terms according to enrichment or depletion
d <- ifelse(sort(scOE_GSEgo_log$NES) > 0,
            cet_pal(2, name='cbd1')[2], cet_pal(2, name='cbd1')[1])
# plot
gsep <- dotplot(scOE_GSEgo_log,
                x='NES', 
                showCategory=number_of_terms,
                label_format=60
                ) +
  ggtitle("Enriched/depleted GO terms in *esg > scute*") +
  scale_colour_gradient(low = cet_pal(3, name='d2')[2],
                        high = cet_pal(3, name='d2')[3]) +
  labs(color = '-log~10~(*p*)') +
  theme(plot.title = element_markdown(),
        plot.title.position = 'panel',
        legend.title = element_markdown(),
        panel.grid = element_blank(),
        axis.text.y = element_text(size=9, colour=d))

suppressWarnings(print(gsep))
```

I think that the signals are not strong enough to give anything meaningful. It may be better to use more focused gene sets ----|>


## 4 Gene set enrichment analysis of custom gene sets


### 4.0 Getting ready


Define a function to import the gene sets in the gmx files
```{r reinventing_wheel_gmx}
import_from_gmx <- function(gmxfile) {
  # read a GMX file and turn it into a df input appropriate for `clusterProfiler`
  df <- read.csv(gmxfile, header=TRUE, sep='\t')
  # remove 'na's in 1st row
  df <- df[ !df[1]=='na', ]
  # pivot
  df <- df %>% pivot_longer(cols = everything(),
                            names_to = 'term',
                            values_to = 'gene')
  # remove rows with no genes
  df <- df %>% subset(gene != '')
  df <- df %>% mutate(gene = str_replace(gene, "FBGN", "FBgn"),
                      term = str_replace(term, ".only", "-only"))
  df <- df %>% mutate(term = str_replace(term, "\\.", " "))
  df <- df %>% mutate(term = str_replace(term, " genes", ""))
  return(df)
}
```


### 4.1 Cell-type specific transcriptomes (Dutta et al., 2015)


First let's run `clusterProfiler::GSEA`:
```{r dutta}
dutta_gmx <- import_from_gmx(paste0(getwd(),'/resources/Dutta.gmx'))

daKD_gse_dutta <- GSEA(geneList=make_degrank(DaKD_deg, mode='log2fc', key='ensemblGeneID'),
                       exponent = 1,
                       minGSSize = 1,
                       maxGSSize = 1000,
                       eps = 0,
                       pvalueCutoff = 1,
                       pAdjustMethod = "BH",
                       TERM2GENE = dutta_gmx,
                       TERM2NAME = NA,
                       verbose = TRUE,
                       seed = FALSE,
                       by = "fgsea")

daOE_gse_dutta <- GSEA(geneList=make_degrank(DaOE_deg, mode='log2fc', key='ensemblGeneID'),
                       exponent = 1,
                       minGSSize = 1,
                       maxGSSize = 1000,
                       eps = 0,
                       pvalueCutoff = 1,
                       pAdjustMethod = "BH",
                       TERM2GENE = dutta_gmx,
                       TERM2NAME = NA,
                       verbose = TRUE,
                       seed = FALSE,
                       by = "fgsea")

dadaOE_gse_dutta <- GSEA(geneList=make_degrank(DaDaOE_deg, mode='log2fc', key='ensemblGeneID'),
                       exponent = 1,
                       minGSSize = 1,
                       maxGSSize = 1000,
                       eps = 0,
                       pvalueCutoff = 1,
                       pAdjustMethod = "BH",
                       TERM2GENE = dutta_gmx,
                       TERM2NAME = NA,
                       verbose = TRUE,
                       seed = FALSE,
                       by = "fgsea")

scOE_gse_dutta <- GSEA(geneList=make_degrank(ScOE_deg, mode='log2fc', key='ensemblGeneID'),
                       exponent = 1,
                       minGSSize = 1,
                       maxGSSize = 1000,
                       eps = 0,
                       pvalueCutoff = 1,
                       pAdjustMethod = "BH",
                       TERM2GENE = dutta_gmx,
                       TERM2NAME = NA,
                       verbose = TRUE,
                       seed = FALSE,
                       by = "fgsea")
```


#### GSEA plots


Now I need to make GSEA plots from the results. I will use `genekitr` for this.
```{r genekitr_customisation}
# the gsea objects generated from clusterProfiler are:
# daKD_gse_dutta, daOE_gse_dutta, dadaOE_gse_dutta, scOE_gse_dutta
daKD_gse_dutta@organism <- 'dm'
daOE_gse_dutta@organism <- 'dm'
dadaOE_gse_dutta@organism <- 'dm'
scOE_gse_dutta@organism <- 'dm'
daKD_gse_dutta_gtr   <- importCP(daKD_gse_dutta, type = 'gsea')
daOE_gse_dutta_gtr   <- importCP(daOE_gse_dutta, type = 'gsea')
dadaOE_gse_dutta_gtr <- importCP(dadaOE_gse_dutta, type = 'gsea')
scOE_gse_dutta_gtr   <- importCP(scOE_gse_dutta, type = 'gsea')

plotGSEA(scOE_gse_dutta_gtr,
         plot_type = "classic",
         show_pathway = scOE_gse_dutta_gtr$gsea_df$ID[1:2],
         #show_gene = scOE_gse_dutta_gtr$genelist[seq(1, by=10, length=2),'ID']
         show_gene = c('da', 'sc', 'emc', 'pros', 'poxn', 'Myo31DF', 'ck', 'nub', 'pdm2', 'esg', 'Dl'))
```


#### Layered heatmaps to display NES across conditions and sets

##### Layered heatmaps: prepare the data

Prepare data to plot heatmaps with NES and padj with the function below, `gseCP_summarise`. This is likely more complicated than it needs to be, as (now) all the GSEA objects have the same number of gene sets, but that is because I am running `GSEA` with `pvalueCutoff = 1`. If you run it with `pvalueCutoff<1`, non-significant gene sets do not even show up in the `@result` slot. In case this is happening, `gseCP_summarise` does a full join.
You run this separately for the NES or adjusted _p_ values for "historical" reasons (I didn't think it through, and had to restore to new syntax that now I don't want to throw away :).
```{r gseCP_summarise}
## extract data from a list of gse objects from clusterProfiler into a single df
## gmx is a df containing the data of a GMX file
## gseCP_list is a list of clusterProfiler GSEA objects
## conditions is a list with the names of the experimental conditions for those GSEA objects
## dvar (data variable) for "name injection" with <data-masking> tidyverse functions
gseCP_summarise <- function(gmx, gseCP_list, conditions, sets.as.factors, dvar, cluster=FALSE) {
  if ( !(dvar %in% c('NES', 'p.adjust')) ) {
    stop("The argument `dvar` must be one of `NES` and `p.adjust`")
  }
  # create template with all term IDs (clusterProfiler trims those that do not give results)
  empty <- data.frame(rep(0, length(unique(gmx$term))),
                    unique(gmx$term))
  names(empty) <- c(dvar, 'ID')
  # create function to merge `empty` with GSE object data for each term ID
  gsemerge <- function(gse, empty, dvar) {
    return(
      # full join selecting injecting dvar with
      # embracing syntax {{}}
      # bangbang !! operator with as.name function
      # glue syntax "{}"
      # dynamic assignment :=
      full_join(dplyr::select(gse, all_of({{dvar}}), ID), empty, by='ID') %>%
        mutate( "{dvar}" := !!as.name(paste0(dvar,".x")) ) %>%
        dplyr::select( {{dvar}}, ID )
      )
  }
  # purr::map to convert the gse_list from S4 objects to their @result slots 
  gseCP_list <- map(gseCP_list, \(x) x@result)
  # apply gsemerge
  gseCP_list <- lapply( gseCP_list, \(x) gsemerge(x, empty, dvar) )
  # name them to associate conditions with the data
  names(gseCP_list) <- conditions
  df <- gseCP_list %>%
    # change the colnames to condition names
    imap(.x = ., ~ set_names(.x, c(.y, "ID"))) %>%
    # merge them all together
    purrr::reduce(full_join, by='ID') %>%
    # tidy up all dvar values in one col
    pivot_longer(cols=-c('ID'), names_to = "condition", values_to = dvar)
  df$condition <- factor(df$condition, levels = conditions)
  df$ID <- factor(df$ID, levels = sets.as.factors)
  if (cluster & length(sets.as.factors)>2) { # `hclust` must have n >= 2 objects to cluster
    if (dvar != 'NES') warning("Clustering is intended for enrichment scores, not p-values")
    vectors <- df %>%
      pivot_wider(names_from = condition, values_from = {{dvar}}) %>%
      mutate_all(replace_na, 100)
    clust <- hclust(dist(as.matrix(vectors[2:length(gseCP_list)])))
    df$ID <- factor(df$ID, levels=vectors$ID[clust$order])
  } else if (cluster & length(sets.as.factors < 3)) {
    warning("`hclust` must have n >= 2 objects to cluster. NES columns will not be clustered.")
  }
  return(df)
}
```

Now we can apply `gseCP_summarise` to the GSEA results from all conditions:
```{r create_dfs_for_NESheatmap}
gse_list <- list(scOE_gse_dutta, dadaOE_gse_dutta, daOE_gse_dutta, daKD_gse_dutta)
conditions <- c('*scute*', '*da:da*', '*da*', '*da^RNAi^*')
# gse_list and conditions must have the same order
sets.as.factors <- c("ISC-only", "Progenitor", "EB-only", "EE-only",
                     "Absorptive", "EC-only", "Differentiation")
NES.df <- gseCP_summarise(dutta_gmx, gse_list, conditions, sets.as.factors, 'NES')
padj.df <- gseCP_summarise(dutta_gmx, gse_list, conditions, sets.as.factors, 'p.adjust')
```

##### Layered heatmaps: plotting function

And we can use this to plot the results in the way we want:
```{r layer.heatmap}
layer.heatmap <- function(NES.df, padj.df, subt) {
  p <- ggplot(NES.df, aes(x=ID, y=condition)) +
    # plot statistic (NES)
    geom_tile(aes(fill=NES), width=1) +
    scale_fill_gradient2(low = cet_pal(3, name='cbd1')[1],
                         mid = cet_pal(3, name='cbd1')[2],
                         high = cet_pal(3, name='cbd1')[3],
                         midpoint = 0) +
    # plot p-value
    # statistically significant with colour and shape `*`
    geom_point(data=subset(padj.df, p.adjust<0.05),
               aes(x=ID, y=condition, colour=-log10(p.adjust)),
               size=3, shape=8, stroke=1.5, alpha=1) +
               #size=5, shape=23, alpha=1) +
    # non-significant in dark gray and shape `x`
    # geom_point(data=subset(padj.df, p.adjust>0.05),
    #            aes(x=ID, y=condition),
    #            size=3, shape=4, stroke=2, alpha=1, colour='gray50') +
    scale_colour_gradient(low = cet_pal(3, name='d2')[2],
                           high = cet_pal(3, name='d2')[3]) +
    coord_equal() +
    theme_bw() +
    ggtitle("Normalised Enrichment Scores",
            subtitle = subt) +
    scale_x_discrete(labels=toupper(levels(NES.df$ID)),
                     position = "top",
                     expand = expansion(mult = 0, add = 0)) +
    scale_y_discrete(expand = expansion(mult = 0, add = 0)) +
    labs(fill = 'NES',
         colour = '-log~10~(_p.adj_)<br><span style = "font-size:8pt;">_p.adj_<0.05</span>') +
    theme(axis.text.x = element_markdown(angle=25, hjust=0,
                                         face='bold', size=10.5),
          axis.text.y = element_markdown(hjust=1, face='bold', size=12),
          axis.title = element_blank(),
          axis.ticks = element_blank(),
          legend.title = element_markdown(hjust=0.5, vjust=0.75),
          legend.direction = 'horizontal',
          legend.position = 'bottom',
          panel.grid = element_blank(),
          panel.border = element_rect(linewidth = 1))
  return(p)
}
```

##### Layered heatmaps for Dutta et al. (2015) gene sets

```{r}
subt <- "for cell-type specific gene sets (Dutta et al., 2015)"
p <- layer.heatmap(NES.df, padj.df, subt)
p + geom_hline(aes(yintercept=3.5), linewidth = 0.5)
```


### 4.2 RNAi screen phenotypes (Zeng et al., 2015)


To simplify later on the GSEA calls:
```{r zeng}
# common gene set
zeng_gmx <- import_from_gmx(paste0(getwd(),'/resources/Zeng.gmx'))
zeng_gmx$term <- str_replace_all(zeng_gmx$term, c(
  "Excess of.ISC.or.EB" = "Excess ISC/EB",
  "ISC to.EC"           = "ISC to EC",
  "ISC to.EE"           = "ISC to EE",
  "Overproliferation"   = "Over-proliferation" ) )
# common parameters
GSEAparams <- list(exponent = 1, minGSSize = 1, maxGSSize = 1000, eps = 0, pvalueCutoff = 1,
                   pAdjustMethod = "BH", TERM2GENE = zeng_gmx, TERM2NAME = NA, verbose = TRUE,
                   seed = FALSE, by = "fgsea")
```

Now we can simply define the rank file and apply GSEA for each condition:
```{r, message=FALSE}
# da knockdown
zrank <- make_degrank(DaKD_deg, mode='log2fc', key='ensemblGeneID')
daKD_gse_zeng <- do.call(
  GSEA, c(list(geneList=zrank), GSEAparams)
  )
# da overexpression
zrank <- make_degrank(DaOE_deg, mode='log2fc', key='ensemblGeneID')
daOE_gse_zeng <- do.call(
  GSEA, c(list(geneList=zrank), GSEAparams)
  )
# da:da overexpression
zrank <- make_degrank(DaDaOE_deg, mode='log2fc', key='ensemblGeneID')
dadaOE_gse_zeng <- do.call(
  GSEA, c(list(geneList=zrank), GSEAparams)
  )
# scute overexpression
zrank <- make_degrank(ScOE_deg, mode='log2fc', key='ensemblGeneID')
scOE_gse_zeng <- do.call(
  GSEA, c(list(geneList=zrank), GSEAparams)
  )
```


#### Layered heatmaps


First to produce the dataframes for `ggplot2`:
```{r create_dfs_for_NESheatmap}
gse_list <- list(scOE_gse_zeng, dadaOE_gse_zeng, daOE_gse_zeng, daKD_gse_zeng)
# `conditions` were defined further above
sets.as.factors <- c("Excess ISC/EB", "Over-proliferation", "ISC to EE", "ISC to EC", "Large nucleus", "ISC death")
NES.df <- gseCP_summarise(zeng_gmx, gse_list, conditions, sets.as.factors, 'NES')
padj.df <- gseCP_summarise(zeng_gmx, gse_list, conditions, sets.as.factors, 'p.adjust')

subt <- "for RNAi-induced phenotypic class gene sets (Zeng et al., 2015)"
p <- layer.heatmap(NES.df, padj.df, subt)
p + geom_hline(aes(yintercept=3.5), linewidth = 0.5)
```

#### GSEA plots

Prepare the data:
```{r genekitr_custom_use}
daKD_gse_zeng@organism <- 'dm'
daOE_gse_zeng@organism <- 'dm'
dadaOE_gse_zeng@organism <- 'dm'
scOE_gse_zeng@organism <- 'dm'
daKD_gse_zeng_gtr   <- importCP(daKD_gse_zeng, type = 'gsea')
daOE_gse_zeng_gtr   <- importCP(daOE_gse_zeng, type = 'gsea')
dadaOE_gse_zeng_gtr <- importCP(dadaOE_gse_zeng, type = 'gsea')
scOE_gse_zeng_gtr   <- importCP(scOE_gse_zeng, type = 'gsea')
```

Tentative plot:
```{r}
plotGSEA(scOE_gse_zeng_gtr,
         plot_type = "classic",
         show_pathway = scOE_gse_zeng_gtr$gsea_df$ID[3],
         show_gene = c() )
```


### 4.3 Gene List Annotation for Drosophila (Hu et al., 2015)


#### Prepare the data for sub.group analyses


```{r collect_GLAD_data}
gladdir <- paste0(getwd(),'/resources/GLAD')
gladfiles <- paste(gladdir, list.files(gladdir), sep='/')
names(gladfiles) <- lapply (
  gladfiles,
  \(x) str_split(str_split(x, "GLAD_")[[1]][[2]], '_')[[1]][[1]] )
read.add <- function(csv, csvlist) {
  df <- read.csv(csv)
  df$term <- names(which(csvlist==csv))
  return(df)
}
datas <- lapply(gladfiles, \(x) read.add(x, gladfiles))
glad_dataset <- do.call(rbind , c(datas, make.row.names=FALSE))
```

Tidy up the strings for term names:
```{r cleanup_names}
substitutions <- c(' signaling pathway$'='',
                   '\\.'=' ',
                   'Co '='Co-',
                   'Non '='Non-',
                   'RTK| RTK'='',
                   "DNA "="DNA-",
                   "TNF "="TNF-",
                   "TGF "="TGF-",
                   ' and '=' / ',
                   'transcription factor|Transcription factor'='TF',
                   ':'='/',
                   '^HLH'='bHLH',
                   '^Basic'='b',
                   '\\|'='')
glad_dataset <- glad_dataset %>%
  # remove 1-2-1 substitutions above
  mutate(across(c(term, Sub.group, Sub.sub.group),
                \(x) str_replace_all(x, substitutions))) %>%
  # remove trailing spaces
  mutate(across(c(term, Sub.group, Sub.sub.group),
                \(x) str_replace_all(x, '  $| $', ''))) %>%
  # start all strings with Uppercase
  mutate(across(c(term, Sub.group, Sub.sub.group),
                \(x) gsub(x, pattern="(^[[:lower:]])", replacement="\\U\\1", perl=TRUE))) %>%
  # substitute empty strings for NAs
  mutate_at(c('term', 'Sub.group', 'Sub.sub.group'), ~na_if(., ''))
View(glad_dataset %>% select(term, Sub.group, Sub.sub.group) %>% distinct() %>% filter(term=='TF/DNA-binding'))
```

Establish the gene sets in tidy dataframes (subdivisions are in lists):
```{r}
# major groups
glad_gmx <- glad_dataset %>%
  select(term, FBgn) %>%
  rename(gene = FBgn)

# subgroups
refine_glad_by <- function(dataset, col.name) {
  # refined by col.name
  subgrouping <- dataset %>%
  # select terms with subgroups
  filter(!is.na(.data[[col.name]])) %>%
  # select the term and subgroup cols
  select(term, {{col.name}}) %>%
  # get unique pairs of terms/subgroup
  distinct()
  # re-make term/gene cols as per subgroups now
  gmx <- lapply(
    unique(subgrouping$term),
    \(x) dataset %>%
      filter( !is.na(.data[[col.name]]) & term=={{x}} ) %>%
      select({{col.name}}, FBgn) %>%
      rename(gene = FBgn, term = {{col.name}})
    )
  names(gmx) <- unique(subgrouping$term)
  # remove terms that only have one subgroup
  gmx <- gmx[ sapply(gmx, \(x) length(unique(x$term)))>1 ]
  return(gmx)
}

glad_sub_gmx <- refine_glad_by(glad_dataset, 'Sub.group')
glad_sub2_gmx <- refine_glad_by(glad_dataset, 'Sub.sub.group')
```

GSEA calls, major lists:
```{r hu_GLAD}
# common gene set
#old_glad_gmx <- import_from_gmx(paste0(getwd(),'/resources/GLAD.gmx'))
# common parameters
GSEAparams <- list(
  exponent = 1,
  minGSSize = 1,
  maxGSSize = 5000,
  eps = 0,
  pvalueCutoff = 1,
  pAdjustMethod = "BH",
  TERM2NAME = NA,
  verbose = TRUE,
  seed = FALSE,
  by = "fgsea"
  )
```

Now we can simply define the rank file and apply GSEA for each condition:
```{r, message=FALSE}
# da knockdown
rank <- make_degrank(DaKD_deg, mode='log2fc', key='ensemblGeneID')
daKD_gse_glad <- do.call(
  GSEA, c(list(geneList=rank, nPermSimple = 1000, TERM2GENE = glad_gmx), GSEAparams)
  )
```
```{r, message=FALSE}
# da overexpression
rank <- make_degrank(DaOE_deg, mode='log2fc', key='ensemblGeneID')
daOE_gse_glad <- do.call(
  GSEA, c(list(geneList=rank, nPermSimple = 1000, TERM2GENE = glad_gmx), GSEAparams)
  )
```
```{r, message=FALSE}
# da:da overexpression
rank <- make_degrank(DaDaOE_deg, mode='log2fc', key='ensemblGeneID')
dadaOE_gse_glad <- do.call(
  GSEA, c(list(geneList=rank, nPermSimple = 1000, TERM2GENE = glad_gmx), GSEAparams)
  )
```
```{r, message=FALSE}
# scute overexpression
# rank <- make_degrank(ScOE_deg, mode='log2fc', key='ensemblGeneID')
# scOE_gse_glad <- do.call(
#   GSEA, c(list(geneList=rank, nPermSimple = 1000000, TERM2GENE = glad_gmx), GSEAparams)
#   )
# saveRDS(scOE_gse_glad, 'output/scOE_gse_glad.RDS')
scOE_gse_glad <- readRDS('output/scOE_gse_glad.RDS')
```


#### Layered heatmaps, main GLAD terms


First to produce the dataframes for `ggplot2`:
```{r create_dfs_for_NESheatmap, fig.width=16}
gse_list <- list(scOE_gse_glad, dadaOE_gse_glad, daOE_gse_glad, daKD_gse_glad)
# `conditions` were defined further above
sets.as.factors <- unique(glad_gmx$term)
NES.df <- gseCP_summarise(glad_gmx, gse_list, conditions, sets.as.factors, 'NES', cluster = TRUE)
padj.df <- gseCP_summarise(glad_gmx, gse_list, conditions, sets.as.factors, 'p.adjust')

subt <- "for Gene List Annotation for Drosophila gene sets (Hu et al., 2015)"
p <- layer.heatmap(NES.df, padj.df, subt)
p + geom_hline(aes(yintercept=3.5), linewidth = 0.5) +
  theme(plot.margin = margin(r = 100))
```


#### GSEA and layered heatmaps for GLAD sub.terms 


Now let's produce the gsea for the sub-terms of GLAD:
```{r GLAD_gsea_subplots}
subglad_gsea <- function(deg, gmx, perms=1000) {
  g <- lapply(
    1:length(gmx), \(x) do.call(
      GSEA,
      c(list(geneList=make_degrank(deg, mode='log2fc', key='ensemblGeneID'),
             nPermSimple = perms,
             TERM2GENE = gmx[[x]]),
        GSEAparams)
      )
  )
  names(g) <- names(gmx)
  return(g)
}
daKD_gse_gladsub   <- subglad_gsea(DaKD_deg, glad_sub_gmx)
daOE_gse_gladsub   <- subglad_gsea(DaOE_deg, glad_sub_gmx)
dadaOE_gse_gladsub <- subglad_gsea(DaDaOE_deg, glad_sub_gmx)
# scOE_gse_gladsub   <- subglad_gsea(ScOE_deg, glad_sub_gmx, perms=100000)
# saveRDS(scOE_gse_gladsub, 'output/scOE_gse_gladsub.RDS')
scOE_gse_gladsub <- readRDS('output/scOE_gse_gladsub.RDS')
scOE_gse_gladsub['Human Disease'] <- NULL
```

Now it is a matter of looping over the names of the GLAD terms with sub/sub-groups, and gather the NES/p.adj data for each condition and each term subdivision.
For the `'Sub.group'` subdivision:
```{r combine_gsea}
gladsub_gsealists <- list(daKD_gse_gladsub,
                          daOE_gse_gladsub,
                          dadaOE_gse_gladsub,
                          scOE_gse_gladsub)

# 'loop' over the Sub.groups in `glad_sub_gmx`
# collect the corresponding GSEA results for each of the conditions
# gather the NES
NES_gladsub_list <- lapply(
  1:length(glad_sub_gmx),
  \(x) gseCP_summarise(glad_sub_gmx[[x]],
                       # this is needed instead of the simpler `lapply(gladsub_gsealists, '[[', x)`
                       # because `gladsub_gsealists` and `glad_sub(2)_gmx` do not coincide in indexing anymore
                       # this can be corrected by running again the GSEA but scute data take a long time
                       # will do that overnight.
                       lapply(gladsub_gsealists, '[[', names(glad_sub_gmx)[[x]]),
                       conditions,
                       unique(glad_sub_gmx[[x]]$term),
                       'NES',
                       cluster = TRUE)
  )
                       
padj_gladsub_list <- lapply(
  1:length(glad_sub_gmx),
  \(x) gseCP_summarise(glad_sub_gmx[[x]],
                       lapply(gladsub_gsealists, '[[', names(glad_sub_gmx)[[x]]),
                       conditions,
                       unique(glad_sub_gmx[[x]]$term),
                       'p.adjust')
  )

names(NES_gladsub_list) <- names(glad_sub_gmx)
names(padj_gladsub_list) <- names(glad_sub_gmx)
```

And plotting:
```{r}
subt <- paste0("for Gene List Annotation for Drosophila gene subsets (Hu et al., 2015)",
               "\n(subsets of ", names(glad_sub_gmx), ")")
p_list <- lapply(
  1:length(glad_sub_gmx),
  \(x) layer.heatmap(NES_gladsub_list[[x]], padj_gladsub_list[[x]], subt[[x]]) +
    geom_hline(aes(yintercept=3.5), linewidth = 0.5) +
    theme(plot.margin = margin(r = 50))
)
```

And to visualise:
```{r}
p_list[[2]]
```

```{r}
p_list[[1]]
```

```{r}
p_list[[3]]
```



```{r}
p_list[[4]]
```

```{r}
 p_list[[5]]
```

```{r}
p_list[[6]]
```

```{r}
p_list[[7]]
```


#### GSEA and layered heatmaps for GLAD sub.sub.terms 


Run GSEA:
```{r GLAD_gsea_subsubplots}
daKD_gse_gladsub2   <- subglad_gsea(DaKD_deg, glad_sub2_gmx)
daOE_gse_gladsub2   <- subglad_gsea(DaOE_deg, glad_sub2_gmx)
dadaOE_gse_gladsub2 <- subglad_gsea(DaDaOE_deg, glad_sub2_gmx)
scOE_gse_gladsub2   <- subglad_gsea(ScOE_deg, glad_sub2_gmx)
```


```{r}
gladsub2_gsealists <- list(daKD_gse_gladsub2,
                           daOE_gse_gladsub2,
                           dadaOE_gse_gladsub2,
                           scOE_gse_gladsub2)

NES_gladsub2_list <- lapply(
  1:length(glad_sub2_gmx),
  \(x) gseCP_summarise(glad_sub2_gmx[[x]],
                       lapply(gladsub2_gsealists, '[[', names(glad_sub2_gmx)[[x]]),
                       conditions,
                       unique(glad_sub2_gmx[[x]]$term),
                       'NES',
                       cluster = TRUE)
  )
                       
padj_gladsub2_list <- lapply(
  1:length(glad_sub2_gmx),
  \(x) gseCP_summarise(glad_sub2_gmx[[x]],
                       lapply(gladsub2_gsealists, '[[', names(glad_sub2_gmx)[[x]]),
                       conditions,
                       unique(glad_sub2_gmx[[x]]$term),
                       'p.adjust')
  )

names(NES_gladsub2_list) <- names(glad_sub2_gmx)
names(padj_gladsub2_list) <- names(glad_sub2_gmx)
```

And plotting:
```{r}
subt <- paste0("for Gene List Annotation for Drosophila gene sub-subsets (Hu et al., 2015)",
               "\n(subsets of ", names(glad_sub_gmx), ")")
names(subt) <- names(glad_sub_gmx)
p2_list <- lapply(
  names(glad_sub2_gmx),
  \(x) layer.heatmap(NES_gladsub2_list[[x]], padj_gladsub2_list[[x]], subt[[x]]) +
    geom_hline(aes(yintercept=3.5), linewidth = 0.5) +
    theme(plot.margin = margin(r = 50))
)
```

```{r}
p2_list[[1]]
```

```{r}
p2_list[[2]]
```

```{r, fig.width=25}
p2_list[[3]]
```

```{r fig.width=20}
p2_list[[4]]
```

NAs are probably due to small groups where the members are simply not detected in the DEG set.


#### GSEA plots

##### For sub-sub-terms

```{r genekitr_customisation_usage1}
for (x in 1:length(daKD_gse_gladsub2)) daKD_gse_gladsub2[[x]]@organism <- 'dm'
for (x in 1:length(daOE_gse_gladsub2)) daOE_gse_gladsub2[[x]]@organism <- 'dm'
for (x in 1:length(dadaOE_gse_gladsub2)) dadaOE_gse_gladsub2[[x]]@organism <- 'dm'
for (x in 1:length(scOE_gse_gladsub2)) scOE_gse_gladsub2[[x]]@organism <- 'dm'

daKD_gse_gladsub2_gtr <- list(
  importCP(daKD_gse_gladsub2$`Major signaling pathways`, type = 'gsea'),
  importCP(daKD_gse_gladsub2$Matrisome, type = 'gsea'),
  importCP(daKD_gse_gladsub2$`TF/DNA-binding`, type = 'gsea'),
  importCP(daKD_gse_gladsub2$Transporters, type = 'gsea')
  )
names(daKD_gse_gladsub2_gtr) <- names(daKD_gse_gladsub2)

daOE_gse_gladsub2_gtr <- list(
  importCP(daOE_gse_gladsub2$`Major signaling pathways`, type = 'gsea'),
  importCP(daOE_gse_gladsub2$Matrisome, type = 'gsea'),
  importCP(daOE_gse_gladsub2$`TF/DNA-binding`, type = 'gsea'),
  importCP(daOE_gse_gladsub2$Transporters, type = 'gsea')
  )
names(daOE_gse_gladsub2_gtr) <- names(daOE_gse_gladsub2)

dadaOE_gse_gladsub2_gtr <- list(
  importCP(dadaOE_gse_gladsub2$`Major signaling pathways`, type = 'gsea'),
  importCP(dadaOE_gse_gladsub2$Matrisome, type = 'gsea'),
  importCP(dadaOE_gse_gladsub2$`TF/DNA-binding`, type = 'gsea'),
  importCP(dadaOE_gse_gladsub2$Transporters, type = 'gsea')
  )
names(dadaOE_gse_gladsub2_gtr) <- names(dadaOE_gse_gladsub2)

scOE_gse_gladsub2_gtr <- list(
  importCP(scOE_gse_gladsub2$`Major signaling pathways`, type = 'gsea'),
  importCP(scOE_gse_gladsub2$Matrisome, type = 'gsea'),
  importCP(scOE_gse_gladsub2$`TF/DNA-binding`, type = 'gsea'),
  importCP(scOE_gse_gladsub2$Transporters, type = 'gsea')
  )
names(scOE_gse_gladsub2_gtr) <- names(scOE_gse_gladsub2)
```

```{r}
plotGSEA(daKD_gse_gladsub2_gtr[[1]],
         plot_type = "classic",
         show_pathway = daKD_gse_gladsub2_gtr[[1]]$gsea_df$ID,
         show_gene = c('da', 'sc', 'emc', 'pros', 'esg', 'Dl'))
```
```{r}
plotGSEA(scOE_gse_gladsub2_gtr[[1]],
         plot_type = "classic",
         show_pathway = scOE_gse_gladsub2_gtr[[1]]$gsea_df$ID,
         show_gene = c('da', 'sc', 'emc', 'pros', 'esg', 'Dl', 'N', 'H', 'Su(H)', 'E(spl)malpha-HLH'))
```

##### For sub-terms

```{r genekitr_customisation_usage1}
for (x in 1:length(daKD_gse_gladsub)) daKD_gse_gladsub[[x]]@organism <- 'dm'
for (x in 1:length(daOE_gse_gladsub)) daOE_gse_gladsub[[x]]@organism <- 'dm'
for (x in 1:length(dadaOE_gse_gladsub)) dadaOE_gse_gladsub[[x]]@organism <- 'dm'
for (x in 1:length(scOE_gse_gladsub)) scOE_gse_gladsub[[x]]@organism <- 'dm'

daKD_gse_gladsub_gtr <- list(
  importCP(daKD_gse_gladsub$Kinases, type = 'gsea'),
  importCP(daKD_gse_gladsub$`Major signaling pathways`, type = 'gsea'),
  importCP(daKD_gse_gladsub$Matrisome, type = 'gsea'),
  importCP(daKD_gse_gladsub$Metabolic, type = 'gsea'),
  importCP(daKD_gse_gladsub$Phosphatases, type = 'gsea'),
  importCP(daKD_gse_gladsub$`TF/DNA-binding`, type = 'gsea'),
  importCP(daKD_gse_gladsub$Transporters, type = 'gsea')
  )
names(daKD_gse_gladsub_gtr) <- names(daKD_gse_gladsub)

daOE_gse_gladsub_gtr <- list(
  importCP(daOE_gse_gladsub$Kinases, type = 'gsea'),
  importCP(daOE_gse_gladsub$`Major signaling pathways`, type = 'gsea'),
  importCP(daOE_gse_gladsub$Matrisome, type = 'gsea'),
  importCP(daOE_gse_gladsub$Metabolic, type = 'gsea'),
  importCP(daOE_gse_gladsub$Phosphatases, type = 'gsea'),
  importCP(daOE_gse_gladsub$`TF/DNA-binding`, type = 'gsea'),
  importCP(daOE_gse_gladsub$Transporters, type = 'gsea')
  )
names(daOE_gse_gladsub_gtr) <- names(daOE_gse_gladsub)

dadaOE_gse_gladsub_gtr <- list(
  importCP(dadaOE_gse_gladsub$Kinases, type = 'gsea'),
  importCP(dadaOE_gse_gladsub$`Major signaling pathways`, type = 'gsea'),
  importCP(dadaOE_gse_gladsub$Matrisome, type = 'gsea'),
  importCP(dadaOE_gse_gladsub$Metabolic, type = 'gsea'),
  importCP(dadaOE_gse_gladsub$Phosphatases, type = 'gsea'),
  importCP(dadaOE_gse_gladsub$`TF/DNA-binding`, type = 'gsea'),
  importCP(dadaOE_gse_gladsub$Transporters, type = 'gsea')
  )
names(dadaOE_gse_gladsub_gtr) <- names(dadaOE_gse_gladsub)

scOE_gse_gladsub_gtr <- list(
  importCP(scOE_gse_gladsub$Kinases, type = 'gsea'),
  importCP(scOE_gse_gladsub$`Major signaling pathways`, type = 'gsea'),
  importCP(scOE_gse_gladsub$Matrisome, type = 'gsea'),
  importCP(scOE_gse_gladsub$Metabolic, type = 'gsea'),
  importCP(scOE_gse_gladsub$Phosphatases, type = 'gsea'),
  importCP(scOE_gse_gladsub$`TF/DNA-binding`, type = 'gsea'),
  importCP(scOE_gse_gladsub$Transporters, type = 'gsea')
  )
names(scOE_gse_gladsub_gtr) <- names(scOE_gse_gladsub)
```

```{r}
plotGSEA(daKD_gse_gladsub_gtr[[1]],
         plot_type = "classic",
         show_pathway = daKD_gse_gladsub_gtr[[1]]$gsea_df$ID,
         show_gene = c('da', 'sc', 'emc', 'pros', 'esg', 'Dl'))
```
```{r}
plotGSEA(scOE_gse_gladsub_gtr[[1]],
         plot_type = "classic",
         show_pathway = scOE_gse_gladsub_gtr[[1]]$gsea_df$ID,
         show_gene = c('da', 'sc', 'emc', 'pros', 'esg', 'Dl', 'N', 'H', 'Su(H)', 'E(spl)malpha-HLH'))
```

##### For main terms

```{r genekitr_customisation_usage1}
daKD_gse_glad@organism <- 'dm'
daOE_gse_glad@organism <- 'dm'
dadaOE_gse_glad@organism <- 'dm'
scOE_gse_glad@organism <- 'dm'

daKD_gse_glad_gtr <- importCP(daKD_gse_glad, type = 'gsea')
daOE_gse_glad_gtr <- importCP(daOE_gse_glad, type = 'gsea')
dadaOE_gse_glad_gtr <- importCP(dadaOE_gse_glad, type = 'gsea')
scOE_gse_glad_gtr <- importCP(scOE_gse_glad, type = 'gsea')
```

```{r}
plotGSEA(daKD_gse_glad_gtr,
         plot_type = "classic",
         show_pathway = daKD_gse_glad_gtr$gsea_df$ID[c(2,3,6)],
         show_gene = c('da', 'sc', 'emc', 'pros', 'esg', 'Dl'))
```
```{r}
plotGSEA(scOE_gse_glad_gtr,
         plot_type = "classic",
         show_pathway = scOE_gse_glad_gtr$gsea_df$ID[c(2,3,6)],
         show_gene = c('da', 'sc', 'emc', 'pros', 'esg', 'Dl', 'N', 'H', 'Su(H)', 'E(spl)malpha-HLH'))
```


## 5 Metabolic signatures using KEGG


### 5.1 Prepare data


This gets us the NCBI gene IDs (a.k.a. entrez gene ID):
```{r get_ncbi-geneid_keys}
ensembl <- useEnsembl(biomart = "ENSEMBL_MART_ENSEMBL",
                     dataset="dmelanogaster_gene_ensembl",
                     host = "https://oct2022.archive.ensembl.org")
attributes <- listAttributes(ensembl)
ezlist <- getBM(attributes=c('entrezgene_id', 'ensembl_gene_id', 'external_gene_name'), mart = ensembl)
ezlist <- drop_na(ezlist)
```

This gets us the KEGG sets for _Drosophila_:
```{r get_kegg_sets}
kegg.sets.dme <- kegg.gsets(species='dme', id.type='entrez')
```

Let us now separate signaling and metabolic pathways (easier to plot later)
```{r tidy_kegg_sets_sig}
# separate sig.naling vs met.abolic pathways and tidy
kegg.sig_tidy <- NULL
KSIG <- kegg.sets.dme$kg.sets[kegg.sets.dme$sig.idx]
for ( x in 1:length(KSIG) ) {
  df <- data.frame(gene = KSIG[[x]])
  df$term <- names(KSIG)[[x]]
  kegg.sig_tidy <- rbind(kegg.sig_tidy, dplyr::select(df, term, gene))
}
# split term names into KEGG ID and description
kegg.sig_tidy <- kegg.sig_tidy %>%
  separate_wider_delim(term, ' ', names = c('term', 'description'),
                       too_many = 'merge')
# show
kegg.sig_tidy[as.integer(seq(from=1, to=nrow(kegg.sig_tidy), length.out = 10)),]
```

Now metabolic ones:
Let us now separate signaling and metabolic pathways (easier to plot later)
```{r tidy_kegg_sets_met}
# separate sig.naling vs met.abolic pathways and tidy
kegg.met_tidy <- NULL
KMET <- kegg.sets.dme$kg.sets[kegg.sets.dme$met.idx]
for ( x in 1:length(KMET) ) {
  df <- data.frame(gene = KMET[[x]])
  df$term <- names(KMET)[[x]]
  kegg.met_tidy <- rbind(kegg.met_tidy, dplyr::select(df, term, gene))
}
# split term names into KEGG ID and description
kegg.met_tidy <- kegg.met_tidy %>%
  separate_wider_delim(term, ' ', names = c('term', 'description'),
                       too_many = 'merge')
kegg.met_tidy[as.integer(seq(from=1, to=nrow(kegg.met_tidy), length.out = 10)),]
```


### 5.2 Signatures


To simplify the calls:
```{r}
# common parameters
kegg.sig_gmx <- kegg.sig_tidy %>% dplyr::select(description, gene)
names(kegg.sig_gmx) <- c('term', 'gene')
GSEAparams_sig <- list(exponent = 1, minGSSize = 1, maxGSSize = 5000, eps = 0, pvalueCutoff = 1,
                       pAdjustMethod = "BH", TERM2GENE = kegg.sig_gmx, TERM2NAME = NA, verbose = TRUE,
                       seed = FALSE, by = "fgsea")

kegg.met_gmx <- kegg.met_tidy %>% dplyr::select(description, gene)
names(kegg.met_gmx) <- c('term', 'gene')
GSEAparams_met <- list(exponent = 1, minGSSize = 1, maxGSSize = 5000, eps = 0, pvalueCutoff = 1,
                       pAdjustMethod = "BH", TERM2GENE = kegg.met_gmx, TERM2NAME = NA, verbose = TRUE,
                       seed = FALSE, by = "fgsea")
```

Now we can do, for each condition with DEG data:
```{r, message=FALSE}
# create rank with ncbi-geneid keys instead of gene symbols
daKD_kegg_rank <- daKD_rank 
names(daKD_kegg_rank) <- ezlist[match(names(daKD_rank), ezlist$external_gene_name), 'entrezgene_id']
# some names will be NA because they are not in the NCBI db (?), so to filter them:
daKD_kegg_rank <- daKD_kegg_rank[-na.action(na.omit(names(daKD_kegg_rank)))]
# now we can do the GSEA safely
daKD_gse_ksig   <- do.call(GSEA, c(list(geneList=daKD_kegg_rank),   GSEAparams_sig))
daKD_gse_kmet   <- do.call(GSEA, c(list(geneList=daKD_kegg_rank),   GSEAparams_met))

# prep
daOE_kegg_rank <- daOE_rank 
names(daOE_kegg_rank) <- ezlist[match(names(daOE_rank), ezlist$external_gene_name), 'entrezgene_id']
daOE_kegg_rank <- daOE_kegg_rank[-na.action(na.omit(names(daOE_kegg_rank)))]
# GSEA
daOE_gse_ksig   <- do.call(GSEA, c(list(geneList=daOE_kegg_rank),   GSEAparams_sig))
daOE_gse_kmet   <- do.call(GSEA, c(list(geneList=daOE_kegg_rank),   GSEAparams_met))

#prep
dadaOE_kegg_rank <- dadaOE_rank 
names(dadaOE_kegg_rank) <- ezlist[match(names(dadaOE_rank), ezlist$external_gene_name), 'entrezgene_id']
dadaOE_kegg_rank <- dadaOE_kegg_rank[-na.action(na.omit(names(dadaOE_kegg_rank)))]
# GSEA
dadaOE_gse_ksig <- do.call(GSEA, c(list(geneList=dadaOE_kegg_rank), GSEAparams_sig))
dadaOE_gse_kmet <- do.call(GSEA, c(list(geneList=dadaOE_kegg_rank), GSEAparams_met))

#prep
scOE_kegg_rank <- scOE_rank 
names(scOE_kegg_rank) <- ezlist[match(names(scOE_rank), ezlist$external_gene_name), 'entrezgene_id']
scOE_kegg_rank <- scOE_kegg_rank[-na.action(na.omit(names(scOE_kegg_rank)))]
# GSEA
scOE_gse_ksig   <- do.call(GSEA, c(list(geneList=scOE_kegg_rank),   GSEAparams_sig))
scOE_gse_kmet   <- do.call(GSEA, c(list(geneList=scOE_kegg_rank),   GSEAparams_met))
```

#### Layered heatmaps

First to produce the dataframes for `ggplot2`, for _signaling_ pathways:
```{r create_dfs_for_NESheatmap, fig.width=100}
gse_list <- list(scOE_gse_ksig, dadaOE_gse_ksig, daOE_gse_ksig, daKD_gse_ksig)
# `conditions` were defined further above
sets.as.factors <- unique(kegg.sig_tidy$description)
NES.df <- gseCP_summarise(kegg.sig_gmx, gse_list, conditions, sets.as.factors, 'NES', cluster=TRUE)
padj.df <- gseCP_summarise(kegg.sig_gmx, gse_list, conditions, sets.as.factors, 'p.adjust')

subt <- "for KEGG *signaling* pathways"
p <- layer.heatmap(NES.df, padj.df, subt)
p + geom_hline(aes(yintercept=3.5), linewidth = 0.5)
```

Now _metabolic_ pathways:
```{r create_dfs_for_NESheatmap, fig.width=100}
gse_list <- list(scOE_gse_kmet, dadaOE_gse_kmet, daOE_gse_kmet, daKD_gse_kmet)
# `conditions` were defined further above
sets.as.factors <- unique(kegg.met_tidy$description)
NES.df <- gseCP_summarise(kegg.met_gmx, gse_list, conditions, sets.as.factors, 'NES', cluster=TRUE)
padj.df <- gseCP_summarise(kegg.met_gmx, gse_list, conditions, sets.as.factors, 'p.adjust')

subt <- "for KEGG *metabolic* pathways"
p <- layer.heatmap(NES.df, padj.df, subt)
p + geom_hline(aes(yintercept=3.5), linewidth = 0.5)
```


# NEXT STEPS

- create filter option for gseCPsummarise so that one can remove conditions where no NES is significant.
- do some plots of individual pathways
- move on to RcisTarget


### 5.3 Plotting


To plot the Path Views as PDF (and if that is not available, as PNG)
```{r create_handled_pathview}
handled_pathview <- function(params) {
  withCallingHandlers(
    do.call ( pathview, c(params, kegg.native = FALSE) ),
    message = function(m) {
      if( length(grep('Try \"kegg.native=T\" instead!', m$message))==1) {
        cat('OK, I will try that\n')
        do.call ( pathview, c(params, kegg.native = TRUE) )
        cat('now it is done with native KEGG PNG\n>>>>> Ignore what comes below: <<<<<\n')
      }
    }
  )
}

# fixed parameters
fixed <- list(
  low = list(gene = cet_pal(3, name='cbd1')[1], cpd = "blue"),
  mid = list(gene = cet_pal(3, name='cbd1')[2], cpd = "gray"),
  high = list(gene = cet_pal(3, name='cbd1')[3], cpd = "yellow"),
  species = "dme",
  kegg.dir = figdir,
  new.signature=FALSE,
  res = 600)
```



```{r}
params <- c(
  list(gene.data = daKD_rank,
       pathway.id = "dme03008",
       out.suffix = "dme03008",
       gene.idtype = 'SYMBOL',
       # limit = list(gene=max(abs(daKD_rank)), cpd=1)
       limit = list(gene=5, cpd=1)
       ),
  fixed)
handled_pathview(params)
```

