---
title: "Analysis of DESeq2 results with TFBD motif enrichment"
description: "DEG analysis based on DESeq2 and RcisTarget"
principal investigator: "Joaquín de Navascués"
researcher: "Joaquín de Navascués"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: 'hide'
    theme: readable
    df_print: paged
---
```{r set-publication-theme, echo=FALSE, cache=FALSE}
ggplot2::theme_set(ggpubr::theme_pubr(base_size=10))
```

```{r setup, echo = FALSE, cache = FALSE}
knitr::opts_chunk$set(dev = 'png', 
                      fig.align = 'center', fig.height = 7, fig.width = 8.5, 
                      pdf.options(encoding = "ISOLatin9.enc"),
                      fig.path='output/figures/', warning=FALSE, message=FALSE)
```

**Libraries needed:**
```{r load_libraries, warning=FALSE, echo=FALSE}
if (!require("librarian")) install.packages("librarian")
# data
librarian::shelf(
  # data manip.
  dplyr, stringr, tidyr, RcisTarget, data.table, DT,
  # plotting
  ggtheme, ggtext, ggrepel, RColorBrewer, cetcolor, hrbrthemes, #?
# convenience
  here)

if(!exists("gseCP_summarise", mode="function")) source("utils.R")
```

**Set working directory where the script is**
```{r setwd}
if (Sys.getenv("USER")=="JQ") {
  setwd("/Users/JQ/Documents/_CODE REPOS/GitHub/Da_RNAseq")
} else if (Sys.getenv("RSTUDIO")==1) {
  setwd( dirname(rstudioapi::getSourceEditorContext(id = NULL)$path) ) # gets what is in the editor
} else {
  setwd(here::here())
  d <- str_split(getwd(),'/')[[1]][length(str_split(getwd(),'/')[[1]])]
  if (d != 'Da_RNAseq') { stop(
    paste0("Could not set working directory automatically to where this",
           " script resides.\nPlease do `setwd()` manually"))
    }
}
getwd()
```

**Path to definitive images (outside repo):**
```{r define_dir2figs}
figdir <- paste0(c(head(str_split(getwd(),'/')[[1]],-1),
                   paste0(tail(str_split(getwd(),'/')[[1]],1), '_figures')),
                 collapse='/')
dir.create(figdir, showWarnings = FALSE)
```


# Analysis of DESeq2 results with RcisTarget


## 1 Load the data


### 1.1 DEG data


This gets us the DGE data from `DESeq2`, identified by FlyBase/Ensembl ID and gene symbol:
```{r load_DEG_data}
# DEG data
DaKD_deg <- readRDS('output/Control_vs_DaKD.RDS')
DaOE_deg <- readRDS('output/Control_vs_DaOE.RDS')
DaDaOE_deg <- readRDS('output/Control_vs_DaDaOE.RDS')
ScOE_deg <- readRDS('output/Control_vs_ScOE.RDS')
head(
  ScOE_deg %>% dplyr::select(gene_symbol, ensemblGeneID, baseMean, log2FoldChange, padj),
  3
)
```

Now we can define the gene sets as up/down-regulated with our _UAS-x_ transgenes:
```{r}
# gene list by condition and up/downregulated
geneLists <- list(
  daRNAi_dn = make_degset(DaKD_deg, up=FALSE, fc_thresh=1.5)$gene_symbol,
  daRNAi_up = make_degset(DaKD_deg, up=TRUE, fc_thresh=1.5)$gene_symbol,
  daOE_dn = make_degset(DaOE_deg, up=FALSE, fc_thresh=1.5)$gene_symbol,
  daOE_up = make_degset(DaOE_deg, up=TRUE, fc_thresh=1.5)$gene_symbol,
  dadaOE_dn = make_degset(DaDaOE_deg, up=FALSE, fc_thresh=1.5)$gene_symbol,
  dadaOE_up = make_degset(DaDaOE_deg, up=TRUE, fc_thresh=1.5)$gene_symbol,
  scOE_dn = make_degset(ScOE_deg, up=FALSE, fc_thresh=1.5)$gene_symbol,
  scOE_up = make_degset(ScOE_deg, up=TRUE, fc_thresh=1.5)$gene_symbol
)
head(geneLists$daRNAi_dn)
```


### 1.2 Motif databases


Now we need the motif Rankings and Annotations, from the [Stein Aerts' lab website](https://resources.aertslab.org/cistarget/).
Matching them correctly is a bit confusing to me, so I first look carefully at the different options:
I download for _D. melanogaster_ release 6.02:

- VERSION 8 (NEW)

  - [Motif rankings `.feather` file v8](https://resources.aertslab.org/cistarget/databases/drosophila_melanogaster/dm6/flybase_r6.02/mc8nr/gene_based/dm6-5kb-upstream-full-tx-11species.mc8nr.genes_vs_motifs.rankings.feather) (782Mb) 
  - [Motif annotation table file v8](https://resources.aertslab.org/cistarget/motif2tf/motifs-v8-nr.flybase-m0.001-o0.0.tbl) (40Mb)

- VERSION 10

  - [Motif rankings `.feather` file v10](https://resources.aertslab.org/cistarget/databases/drosophila_melanogaster/dm6/flybase_r6.02/mc_v10_clust/gene_based/dm6_v10_clust.genes_vs_motifs.rankings.feather) (47Mb) 
  - [Motif annotation table file v10](https://resources.aertslab.org/cistarget/motif2tf/motifs-v10nr_clust-nr.flybase-m0.001-o0.0.tbl) (69Mb)


Load the different motif Annotations (mA) and motif Rankings (mR)
```{r}
rkURLbase <- paste0('https://resources.aertslab.org/cistarget/databases/',
                   'drosophila_melanogaster/dm6/flybase_r6.02')
rk_v8_url <- paste0(rkURLbase, '/mc8nr/gene_based/',
                   'dm6-5kb-upstream-full-tx-11species.mc8nr.genes_vs_motifs.rankings.feather')
an_v8_url <- paste0('https://resources.aertslab.org/cistarget/motif2tf/',
                    'motifs-v8-nr.flybase-m0.001-o0.0.tbl')
rk_v10_url <- paste0(rkURLbase, '/mc_v10_clust/gene_based/',
                     'dm6_v10_clust.genes_vs_motifs.rankings.feather')
an_v10_url <- paste0('https://resources.aertslab.org/cistarget/motif2tf/',
                     'motifs-v10nr_clust-nr.flybase-m0.001-o0.0.tbl')

# establish the resource sub-directory
motifpath <- paste0(getwd(), "/resources/motifdbs")
if (!dir.exists( motifpath )) dir.create(motifpath)
# read (or download, if the first time) the rankings and annotations 
```

```{r}
tryCatch(
  { motif_annotations_dm6_v8 <- readRDS(paste0(motifpath, "/motif_annotations_dm6_v8.RDS")) },
  error = function(e) {
    message(e)
    message('\n"motif_annotations_dm6_v8" is not available locally -- it will be downloaded and saved.')
    motif_annotations_dm6_v8 <- importAnnotations(an_v8_url)
    saveRDS(motif_annotations_dm6_v8, paste0(motifpath, "/motif_annotations_dm6_v8.RDS"))
    }
)
tryCatch(
  { options(timeout = max(600, getOption("timeout")))
    motif_rankings_dm6_v8 <- readRDS(paste0(motifpath, "/motif_rankings_dm6_v8.RDS")) },
  error = function(e) {
    message('\n"motif_rankings_dm6_v8" is not available locally -- it will be downloaded and saved.\n\n')
    message(e)
    # timeout is in seconds, and may depend on the connection:
    options(timeout = max(1500, getOption("timeout")))
    # tried tempfile(), didn't work
    destfile <- paste0(motifpath, 'motif_rankings_dm6_v8.feather')
    download.file(url = rk_v8_url,
                  destfile = destfile,
                  mode='wget')
    motif_rankings_dm6_v8 <- importRankings(destfile)
    file.remove(destfile)
    saveRDS(motif_rankings_dm6_v8, paste0(motifpath, "/motif_rankings_dm6_v8.RDS"))
    }
)
tryCatch(
  { motif_annotations_dm6_v10 <- readRDS(paste0(motifpath, "/motif_annotations_dm6_v10.RDS")) },
  error = function(e) {
    message(e)
    message('\n"motif_annotations_dm6_v10" is not available locally -- it will be downloaded and saved.')
    motif_annotations_dm6_v10 <- importAnnotations(an_v10_url)
    saveRDS(motif_annotations_dm6_v10, paste0(motifpath, "/motif_annotations_dm6_v10.RDS"))
    }
)
tryCatch(
  { motif_rankings_dm6_v10 <- readRDS(paste0(motifpath, "/motif_rankings_dm6_v10.RDS")) },
  error = function(e) {
    message('\n"motif_rankings_dm6_v10.RDS" is not available locally -- it will be downloaded and saved.\n\n')
    message(e)
    options(timeout = max(600, getOption("timeout"))) # number is in seconds, and may depend on the connection
    # tried tempfile(), didn't work
    destfile <- paste0(motifpath, 'motif_rankings_dm6_v10.feather')
    download.file(url = rk_v10_url,
                  destfile = destfile,
                  mode='wget')
    motif_rankings_dm6_v10 <- importRankings(destfile)
    file.remove(destfile)
    saveRDS(motif_rankings_dm6_v10, paste0(motifpath, "/motif_rankings_dm6_v10.RDS"))
    }
)
# place last column first, as RcisTarget assumes that is the position of the motif names
motif_rankings_dm6_v10@rankings <- dplyr::relocate(motif_rankings_dm6_v10@rankings, motifs)
motif_rankings_dm6_v8@rankings <- dplyr::relocate(motif_rankings_dm6_v8@rankings, motifs)
```


## 2 Perform motif enrichment with `RcisTarget`


### 2.1 Run with the two versions of the motif annotations


Now I need to see how many motifs are there per database:
```{r}
motif.table <- data.frame(
  motif_annot = c(length(unique(motif_annotations_dm6_v8$motif)), length(unique(motif_annotations_dm6_v10$motif))),
  motif_ranks = c(nrow(motif_rankings_dm6_v8), nrow(motif_rankings_dm6_v10))
)
rownames(motif.table) <- c('8new', '10cls')
motif.table
```

So it seems that the combination with more possibilities is to match the v8 motif ranks with the v10 annotation -- the website recommends the opposite, but it is also true that this is in the context of the pySCENE pipeline, which is to build networks with scRNAseq data.

Let's see what we get with a simple comparison:
```{r}
tryCatch(
  { motrich_pure8 <- readRDS(paste0(getwd(), "/output", "/motrich_pure8.RDS")) },
  error = function(e) {
    motrich_pure8 <- cisTarget(geneLists, motif_rankings_dm6_v8, motifAnnot = motif_annotations_dm6_v8)
    saveRDS(motrich_pure8, paste0(getwd(), "/output", "/motrich_pure8.RDS"))
    }
)
tryCatch(
  { motrich_hybr8 <- readRDS(paste0(getwd(), "/output", "/motrich_hybr8.RDS")) },
  error = function(e) {
    motrich_hybr8 <- cisTarget(geneLists, motif_rankings_dm6_v8, motifAnnot = motif_annotations_dm6_v10)
    saveRDS(motrich_hybr8, paste0(getwd(), "/output", "/motrich_hybr8.RDS"))
    }
)
```


### 2.2 Merge the association with TFs


On inspection of the results, most of the information is contained in the 'pure v8' dataframe, whereas the 'hybrid v8-v10' is mostly redundant - but not fully. So I will simply merge them (the connection between motifs and TFs should not depend on how the databases were assembled and the criteria for inclusion is the same).
```{r}
motrich <- motrich_pure8 %>%
  # select only the data I am interested in from the pure v8 enrichment analysis
  dplyr::select(-c(TF_lowConf, rankAtMax, nEnrGenes)) %>%
  # add the data from the hybrid v8-v10 analysis
  tibble::add_column(TF_highConf2 = motrich_hybr8$TF_highConf) %>%
  # remove trailing '. 's and remove source evidence clarifications between brackets
  mutate( across(contains('TF_highConf'), ~ str_remove_all(., " \\(.+\\)|\\. |\\.")) ) %>%
  mutate( across(contains('TF_highConf'), ~ str_remove(., " $")) ) %>%
  # merge v8 / v10 annotations
  unite('TFs', contains('TF_highConf'), sep = "; ") %>%
  # remove piloting/trailing separators
  mutate(TFs = str_remove(TFs, "^; |; $")) %>%
  # split TF names in individual strings
  mutate(TFs = str_split(TFs, '; '))
# remove duplicate TFs
motrich$TFs <- lapply(motrich$TFs, unique)
head(motrich)
```


### 2.3 Associate TFs with TF class using FlyBase's Gene Groups


#### Download the Gene Groups

To assign to gene families, let's start by obtaining the gene family mapping from FlyBase:
```{r, message=FALSE}
geneg.source <- 'http://ftp.flybase.org/releases/FB2023_02/precomputed_files/genes/gene_group_data_fb_2023_02.tsv.gz'
geneg <- data.table::fread(geneg.source)
names(geneg)[[1]] <- 'FB_group_id'
geneg <- geneg %>% dplyr::select( !contains('_id') )
names(geneg) <- c('group_symbol', 'group_name', 'parent_symbol', 'gene_symbol')
nrow(geneg); head(geneg)
```

Now we have a look at how the TFs in `geneg` and `motrich` match each other.

#### Clean up the Gene Group list

Let's start with how many TFs whose motifs were detected by `RcisTarget` do not have a Gene Group:
```{r}
# unique TF names excluding "" for motifs without a _Drosophila_ TF match
detected.tfs <- unique(unlist(motrich$TFs))
detected.tfs <- detected.tfs[ nchar(detected.tfs)>0 ]
# TFs detected in RcisTarget but not present in the group member db:
detected.tfs[!(detected.tfs %in% geneg$gene_symbol)]
```

Of these:

- _h_ is the usual name for _hairy_, which in FB (and therefore in `geneg`) is called _hry_ `-|>` change name in `geneg`
- _CG14440_ and _CG14442_ are homologs of human _ZNF821_, a ZF-C2H2 factor (DIOPT) `-|>` add annotation to `geneg`
- _ocm_ is homolog of human _TBX22_, a T-BOX factor (DIOPT) `-|>` add annotation to `geneg`
- _NK71_ I can't find it, except in this db from the Aerts'lab and poster from the Stark's lab `-|>` remove from `motrich`

```{r}
# add needed gene group annotations
geneg <- geneg %>%
  # change the symbol for _hairy_ from FB official 'hry' to standard 'h'
  mutate(gene_symbol = str_replace(gene_symbol, "^hry$", "h")) %>%
  # add ocm as a T-BOX gene
  add_row(
    geneg %>%
      filter(str_detect(group_name, "T-BOX")) %>%
      summarise(first(.)) %>%
      mutate(gene_symbol='ocm')
    ) %>%
  # add CG14440 and CG14442 as ZF-C2H2 genes
  bind_rows(
    geneg %>%
      filter(str_detect(group_name, "C2H2")) %>%
      head(2) %>%
      mutate(gene_symbol=c('CG14440', 'CG14442'))
    )
# now get rid of all genes that have not been recovered by RcisTarget
geneg <- geneg %>%
  filter(gene_symbol %in% detected.tfs)

# to remove NK71 from the `motrich` TFs, it is better to retrace our steps:
motrich$TFs <- lapply(motrich$TFs, \(x) str_remove(x, 'NK71'))
```

Now we can have a look at genes that have multiple matches in the gene groups:
```{r}
# TFs RcisTarget with multiple gene group memberships:
multifam <- lapply(detected.tfs, function(x) nrow(filter(geneg, gene_symbol==x)))>1
geneg %>%
  filter(gene_symbol %in% detected.tfs[multifam]) %>%
  arrange(gene_symbol)
```

To clean this up, we can simply:

- remove the gene group that does not have a parent symbol
- that will leave us with _Abd-B_, _Myb_, _ham_, _pb_, _sr_, _toy_, _zfh1_

Make gene-specific filters:
```{r}
double_parent_genes <- geneg %>%
  filter(gene_symbol %in% detected.tfs[multifam] & parent_symbol!='') %>%
  filter(geneg %>%
           filter(gene_symbol %in% detected.tfs[multifam] & parent_symbol!='') %>%
           dplyr::select(gene_symbol) %>%
           duplicated()
         ) %>%
  dplyr::select(gene_symbol)
double_parent_genes <- sort(unlist(double_parent_genes))
names(double_parent_genes) <- NULL
# discriminate manually ¯\_(ツ)_/¯ which row to keep:
double_parent_symbol <- c("HOX-C", "KMT", "HTH", "HOX-C", "ULT", "HBTF", "ZN-TF")
```

Apply filters:
```{r}
geneg <- geneg %>%
  # remove the group associations that are multiple and without a parent_symbol
  filter( ! (gene_symbol %in% detected.tfs[multifam] & parent_symbol=='') ) %>%
  # remove the specific associations
  filter( !(gene_symbol==double_parent_genes[[1]] & parent_symbol==double_parent_symbol[[1]]) &
          !(gene_symbol==double_parent_genes[[2]] & parent_symbol==double_parent_symbol[[2]]) &
          !(gene_symbol==double_parent_genes[[3]] & parent_symbol==double_parent_symbol[[3]]) &
          !(gene_symbol==double_parent_genes[[4]] & parent_symbol==double_parent_symbol[[4]]) &
          !(gene_symbol==double_parent_genes[[5]] & parent_symbol==double_parent_symbol[[5]]) &
          !(gene_symbol==double_parent_genes[[6]] & parent_symbol==double_parent_symbol[[6]]) &
          !(gene_symbol==double_parent_genes[[7]] & parent_symbol==double_parent_symbol[[7]]) )
```

#### Associate TF class (Gene Group) to TFs from `RcisTarget`

Now we can provide gene group symbols to the TFs from `RcisTarget`, which is more important than the identity of specific TFs.
```{r}
# for every list of TFs matched to a motif...
motrich$TFclass <- lapply(motrich$TFs, \(x)
  # ... get all the different TF classes corresponding to those TFs
  unique( unlist( lapply( x, \(y) filter(geneg, gene_symbol==y)$group_symbol) ) )
  )
```

As some motifs can be bound by different classes of TFs, we have to do:
```{r}
# before, I simply unnested `TFclass` to replicate motifs bound by different TF classes
motrich$TFclass <- lapply(motrich$TFs, \(x)
  # ... get all the TF classes corresponding to those TFs, **allowing repetition**
  unique( unlist( lapply( x, \(y) filter(geneg, gene_symbol==y)$group_symbol) ) )
  )
# however I think this is misleading (e.g. some GATA boxes are also bound by Da, but this makes no sense to highlight)
# So now I will get them all:
motrich$TFclassRep <- lapply(motrich$TFs, \(x)
  # ... get all the TF classes corresponding to those TFs, **allowing repetition**
  unlist( lapply( x, \(y) filter(geneg, gene_symbol==y)$group_symbol) )
  )
# So now, for every list of TFclasses matched (repeatedly) to a motif...
motrich$TFmainClass <- lapply(motrich$TFclassRep, \(x)
  # ... get the most abundant one
  sort(table(x), decreasing=TRUE)[1] |> names()
  )
# remove NULLs caused by character(0) in $TFs
motrich$TFmainClass[ sapply(motrich$TFmainClass, is.null) ] <- ''
length(unique(motrich$TFmainClass))
```

This is still too much for the plot. So I am going to simplify the list of different TF classes by hand:
```{r}
# to map `geneg` descriptions to better, less abundant names:
name2largeclass <- list(
"FORK HEAD BOX TRANSCRIPTION FACTORS" = 'FOX',
"CBF DOMAIN TRANSCRIPTION FACTOR COMPLEX" = 'CBF',
"TATA-BINDING PROTEIN AND TBP-RELATED FACTORS" = 'TBP/TBP-related',
"NK-LIKE HOMEOBOX TRANSCRIPTION FACTORS" = 'Hbox',
"BASIC LEUCINE ZIPPER TRANSCRIPTION FACTORS" = 'bZIP',
"NUCLEAR RECEPTOR SUBFAMILY 0 (LIGAND-INDEPENDENT) TRANSCRIPTION FACTORS" = "NR",
"HIGH MOBILITY GROUP BOX TRANSCRIPTION FACTORS" = "HMG",
"SINE OCULIS HOMEOBOX TRANSCRIPTION FACTORS" = 'Hbox',
"UNCLASSIFIED DNA BINDING DOMAIN TRANSCRIPTION FACTORS" = 'unclassified',
"TEA DOMAIN TRANSCRIPTION FACTORS" = "TEAD",
"MADS-BOX TRANSCRIPTION FACTORS" = "MADS",
"SIRTUIN LYSINE DEACETYLASES" = 'SIRT',
"WNT ENHANCEOSOME" = 'WNTE',
"T-BOX TRANSCRIPTION FACTORS" = 'TBX',
"NUCLEAR RECEPTOR (LIGAND-DEPENDENT) TRANSCRIPTION FACTORS" = "NR",
"SANT-MYB DOMAIN TRANSCRIPTION FACTORS" = "SANT-MYB",
"C2H2 ZINC FINGER TRANSCRIPTION FACTORS" = 'ZnF',
"GATA TRANSCRIPTION FACTORS" = 'GATA',
"BASIC HELIX-LOOP-HELIX TRANSCRIPTION FACTORS" = 'bHLH',
"E2F TRANSCRIPTION FACTORS" = 'E2F',
"GLIAL CELL MISSING TRANSCRIPTION FACTORS" = "GCM",
"PAIRED-LIKE HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox",
"NUCLEAR FACTOR - KAPPA B" = "NFkB",
"POU HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox",
"CUT HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox",
"HOX-LIKE HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox",
"ZINC FINGER HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox",
"TALE HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox",
"NUCLEAR FACTOR OF ACTIVATED T-CELLS TRANSCRIPTION FACTORS" = "NFAT",
"LIM HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox",
"ETS DOMAIN TRANSCRIPTION FACTORS" = "ETS",
"GATOR2 COMPLEX" = "GAT2",
"PAIRED HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox"
)
# to map `geneg` descriptions with symbols
class2name <- (geneg %>%
  dplyr::select(group_symbol, group_name) %>%
  distinct() %>% dplyr::select(group_name))[[1]]
names(class2name) <- (geneg %>%
  dplyr::select(group_symbol, group_name) %>%
  distinct() %>% dplyr::select(group_symbol))[[1]]
# apply to motrich
## first get the NULLs in TFmainClass
motrich <- motrich %>%
  mutate(TFLargeClass = recode(unlist(TFmainClass), !!!class2name)) %>%
  mutate(TFLargeClass = recode(unlist(TFLargeClass), !!!name2largeclass))
length(unique(motrich$TFLargeClass))
```

Somewhat better, but I would like something like ~6-7 colours at most. So I will merge the groups with less than ~10 hits into one after the next step.


## 3 Plot the results


### 3.1 Beeswarm plot (dotplot)


I want to create a categorical (by experimental condition) 'beeswarm' dot plot that shows the Y position of the dot as the NES value, and the class of TF as colour.

The minimal datasets are:
```{r}
# genes up
mots_up <- motrich %>%
  dplyr::select(c(TFs, geneSet, NES, TFLargeClass)) %>%
  filter(str_detect(geneSet, 'up'))
mots_up$geneSet <- factor(mots_up$geneSet,
                          levels=c("daRNAi_up", "daOE_up", "dadaOE_up", "scOE_up"),
                          labels=c("*da^RNAi^*", "*da*", "*da:da*", "*scute*"))
# genes down
mots_dn <- motrich %>%
  dplyr::select(c(TFs, geneSet, NES, TFLargeClass)) %>%
  filter(str_detect(geneSet, 'dn'))
mots_dn$geneSet <- factor(mots_dn$geneSet,
                          levels=c("daRNAi_dn", "daOE_dn", "dadaOE_dn", "scOE_dn"),
                          labels=c("*da^RNAi^*", "*da*", "*da:da*", "*scute*"))
```

To coarse-grain the TF classes:
```{r}
top5up <- mots_up %>%
  dplyr::count(TFLargeClass) %>%
  # remove '' and unclassified -- the latter have undue weight
  filter(nchar(TFLargeClass)>1 & TFLargeClass!='unclassified') %>%
  arrange(n) %>%
  tail(5) %>%
  select(TFLargeClass) %>% unlist()

mots_up <- mots_up %>%
  mutate(TFcla6 = if_else(TFLargeClass %in% top5up, TFLargeClass, 'Others'))
mots_up$TFcla6 <- factor(mots_up$TFcla6, levels=c(rev(top5up), 'Others'))
indices <- lapply(unique(mots_up$geneSet), \(x) nrow(filter(mots_up, geneSet==x)):1 )
mots_up$index <- unlist(indices)

top5dn <- mots_dn %>%
  dplyr::count(TFLargeClass) %>%
  filter(nchar(TFLargeClass)>1 & TFLargeClass!='unclassified') %>%
  arrange(n) %>%
  tail(5) %>%
  select(TFLargeClass) %>% unlist()

mots_dn <- mots_dn %>%
  mutate(TFcla6 = if_else(TFLargeClass %in% top5dn, TFLargeClass, 'Others'))
mots_dn$TFcla6 <- factor(mots_dn$TFcla6, levels=c(rev(top5dn), 'Others'))
indices <- lapply(unique(mots_dn$geneSet), \(x) nrow(filter(mots_dn, geneSet==x)):1 )
mots_dn$index <- unlist(indices)
```

To plot:
```{r dotplot_NES_TFclass, fig.width=4, fig.height=4}
palette <- c( brewer.pal(12, 'Paired')[c(6, 8, 4, 2, 10)], '#CCCCCC' )
p <- ggplot(mots_up, aes(x=geneSet, y=NES, fill=TFcla6, colour=TFcla6)) + 
  geom_dotplot(binaxis='y',
               stackdir='center',
               stackgroups = TRUE,
               method='histodot',
               binwidth=0.1,
               dotsize=1,
               alpha=1,
               stroke=0) +
  scale_fill_manual(name='',
                    breaks=levels(mots_up$TFcla6),
                    values=palette,
                    labels=levels(mots_up$TFcla6),
                    guide=guide_legend(nrow = 1)) +
  scale_colour_manual(name='',
                    breaks=levels(mots_up$TFcla6),
                    values=palette,
                    labels=levels(mots_up$TFcla6),
                    guide=guide_legend(nrow = 1)) +
  xlab('TFs linked to enriched motifs') +
  theme(axis.text.x = element_markdown(face='bold', size=10),
        aspect.ratio = 0.5,
        legend.position = 'bottom')
p
```

Now we need to identify specific TFs: _da_, members of the _ac/sc_ Complex (_ac_, _sc_, _ase_, _l(1)sc_), and maybe other essential ones (_klu_, _nub/pdm1_)
```{r mark_special_TFs, fig.width=5, fig.height=4}
# using https://stackoverflow.com/questions/44991607
tfs.regex <- c('da', 'sc', 'ase', 'E\\(spl\\)')
tfs.var <- c('da', 'sc', 'ase', 'E(spl)')
for (j in 1:length(tfs.regex)) {
  mots_up <- mots_up %>%
    mutate('{tfs.var[[j]]}' := if_else(
      (str_detect(as.character(mots_up$TFs), tfs.regex[[j]]) & TFcla6=='bHLH'),
      tfs.var[[j]], NA) )
}
mots_up <- mots_up %>%
  unite('tf.labs', all_of(tfs.var), sep = ', ', na.rm = TRUE, remove=FALSE)

# to get the lines of ggrepel text where we want:
built <- ggplot_build(p)
point.pos <- built$data[[1]]
size <- dev.size(units = 'px')
extent <- with(built$layout$panel_params[[1]], abs(c(diff(x.range), diff(y.range))))
bw <- point.pos$binwidth[[1]]
# por la cuenta de la vieja, al final
xtext <- point.pos$x + point.pos$stackpos * bw * (size[2] / size[1]) * (extent[1] / extent[2]) * (38/64)
ytext <- point.pos$y

p +
  geom_text_repel(
    # reindex - point.pos is organised by X values, then by FILL colour
    # which equates to ordering tf.lab as X, then TFclass, then NES
    aes(label = arrange(mots_up, geneSet, TFcla6, NES)$da,
        x = xtext, y = ytext,
        colour = arrange(mots_up, geneSet, TFcla6, NES)$TFcla6),
    size=3,
    # general
    max.overlaps = Inf,
    # position
    direction = 'both',
    nudge_x = -0.25,
    nudge_y = 1.5,
    # segment
    #arrow = arrow(angle=20, length = unit(0.015, "npc")),
    min.segment.length = 0,
    segment.color = 'gray30',
    segment.alpha = 0.75,
    segment.size = 0.2) +
  geom_text_repel(
    # reindex - point.pos is organised by X values, then by FILL colour
    # which equates to ordering tf.lab as X, then TFclass, then NES
    aes(label = arrange(mots_up, geneSet, TFcla6, NES)$sc,
        x = xtext, y = ytext,
        colour = arrange(mots_up, geneSet, TFcla6, NES)$TFcla6),
    size=3,
    # general
    max.overlaps = Inf,
    # position
    direction = 'both',
    nudge_x = 0.4,
    nudge_y = 0,
    # segment
    #arrow = arrow(angle=20, length = unit(0.015, "npc")),
    min.segment.length = 0,
    segment.color = 'gray30',
    segment.alpha = 0.75,
    segment.size = 0.2) +
  guides(fill = guide_legend(nrow=1),
         colour = guide_legend(nrow=1)) +
  ylim(2.5,10)
```

```{r}
ggsave(paste0(figdir, '/motifs_up_TFclass.pdf'))
```

Now for the downregulataed genes:

```{r dotplot_NES_TFclass_down, fig.width=4.5, fig.height=4}
palette <- c( brewer.pal(12, 'Paired')[c(8, 6, 2, 4, 10)], '#CCCCCC' )
q <- ggplot(mots_dn, aes(x=geneSet, y=NES, fill=TFcla6, colour=TFcla6)) + 
  geom_dotplot(binaxis='y',
               stackdir='center',
               stackgroups = TRUE,
               method='histodot',
               binwidth=0.1,
               dotsize=1,
               alpha=1,
               stroke=0) +
  scale_fill_manual(name='',
                    breaks=levels(mots_dn$TFcla6),
                    values=palette,
                    labels=levels(mots_dn$TFcla6),
                    guide=guide_legend(nrow = 1)) +
  scale_colour_manual(name='',
                    breaks=levels(mots_dn$TFcla6),
                    values=palette,
                    labels=levels(mots_dn$TFcla6),
                    guide=guide_legend(nrow = 1)) +
  xlab('TFs linked to enriched motifs') +
  theme(axis.text.x = element_markdown(face='bold', size=10),
        aspect.ratio = 0.5,
        legend.position = 'bottom')

tfs.regex <- c('da', 'sc', 'ase', 'E\\(spl\\)')
tfs.var <- c('da', 'sc', 'ase', 'E(spl)')
for (j in 1:length(tfs.regex)) {
  mots_dn <- mots_dn %>%
    mutate('{tfs.var[[j]]}' := if_else(
      (str_detect(as.character(mots_dn$TFs), tfs.regex[[j]]) & TFcla6=='bHLH'),
      tfs.var[[j]], NA) )
}
mots_dn <- mots_dn %>%
  unite('tf.labs', all_of(tfs.var), sep = ', ', na.rm = TRUE, remove=FALSE)

# to get the lines of ggrepel text where we want:
built <- ggplot_build(q)
point.pos <- built$data[[1]]
size <- dev.size(units = 'px')
extent <- with(built$layout$panel_params[[1]], abs(c(diff(x.range), diff(y.range))))
bw <- point.pos$binwidth[[1]]
# por la cuenta de la vieja, al final
xtext <- point.pos$x + point.pos$stackpos * bw * (size[2] / size[1]) * (extent[1] / extent[2]) * (38/64)
ytext <- point.pos$y

q +
  geom_text_repel(
    # reindex - point.pos is organised by X values, then by FILL colour
    # which equates to ordering tf.lab as X, then TFclass, then NES
    aes(label = arrange(mots_dn, geneSet, TFcla6, NES)$da,
        x = xtext, y = ytext,
        colour = arrange(mots_dn, geneSet, TFcla6, NES)$TFcla6),
    size=2,
    # general
    max.overlaps = Inf,
    # position
    direction = 'both',
    nudge_x = -0.25,
    nudge_y = 0,
    force_pull = 0.5,
    # segment
    #arrow = arrow(angle=20, length = unit(0.015, "npc")),
    min.segment.length = 0,
    segment.color = 'gray30',
    segment.alpha = 0.75,
    segment.size = 0.2) +
  geom_text_repel(
    aes(label = arrange(mots_dn, geneSet, TFcla6, NES)$`E(spl)`,
        x = xtext, y = ytext,
        colour = arrange(mots_dn, geneSet, TFcla6, NES)$TFcla6),
    size=2,
    # general
    max.overlaps = Inf,
    # position
    direction = 'both',
    nudge_x = -0.25,
    nudge_y = 3,
    xlim = c(0.4, 3.5),
    ylim = c(6,10),
    force = 0.8,
    box.padding = 0.2,
    force_pull = 0,
    # segment
    #arrow = arrow(angle=20, length = unit(0.015, "npc")),
    min.segment.length = 0,
    segment.color = 'gray30',
    segment.alpha = 0.75,
    segment.size = 0.2) +
  guides(fill = guide_legend(nrow=1),
         colour = guide_legend(nrow=1)) +
  ylim(2.5,10) +
  theme(plot.margin = margin(r = 50))
```

```{r}
ggsave(paste0(figdir, '/motifs_down_TFclass.pdf'))
```


### 3.1 Rank lollipop plot (experiment)


The beeswarm plot loses a bit the idea of the rank, so I try the lollipop plot, which will show as a continuously increasing function with the colour indicating the type of TF.

```{r}
palette <- c( brewer.pal(12, 'Paired')[c(6, 8, 4, 2, 10)], '#CCCCCC' )
r <- ggplot(mots_up, aes(x=index, y=NES)) +
  geom_segment(aes(x = index, xend = index, y = 0, yend = NES, colour=TFcla6),
               linewidth=0.1) +
  geom_point(aes(fill=TFcla6),
             size = 1, shape=21, stroke=0) +
  scale_colour_manual(name='',
                      breaks=mots_up$TFcla6,
                      values=palette,
                      labels=mots_up$TFcla6) +
  scale_fill_manual(name='',
                    breaks=mots_up$TFcla6,
                    values=palette,
                    labels=mots_up$TFcla6) +
  facet_grid(. ~ geneSet,  scales = "free", space = 'free') +
  xlab("TF class for enriched motif") +
  theme(strip.text.x = element_markdown(face='bold', colour='black', size=10),
        strip.background.x = element_rect(fill=cet_pal(9, name='cbd1')[[6]],
                                          colour='white'),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        legend.position = 'bottom'
        )
r
```

Something is going on with the colours (which didn't happen the other day!), but I will not continue with the experiment, as I do not think it looks any better.

