---
title: "GSEA of RNAseq DGE data"
output:
  html_document:
    toc: yes
    toc_depth: 2
---

```{r}
library(msigdbr)
library(fgsea)
library(ggplot2)
library(dplyr)


m_df = msigdbr(species = "Drosophila melanogaster", category = "C5")

# I want unique rows for Drosophila, and because they have human homolog, Drosophila genes are duplicated. Remove Human data and then find unique
m_df$human_gene_symbol <- NULL
m_df$sources <- NULL

m_df <- unique(m_df)


#Use the gene sets data frame for fgsea.

m_list = m_df %>% split(x = .$gene_symbol, f = .$gs_name)

### Preparation of all sets to analyse ###

setwd('/Users/aleix/Documents/bioinformatics/RNAseq/')
working_dir <- getwd()

# load all datasets obtained with DESEQ2

WT_vs_2D <- read.table(file="output/differential_expression/2D_vs_WT.txt", header=TRUE, sep="\t")
WT_vs_da <- read.table(file="output/differential_expression/da_vs_WT.txt", header=TRUE, sep="\t")
WT_vs_sc <- read.table(file="output/differential_expression/sc_vs_WT.txt", header=TRUE, sep="\t")
gene_list <- read.table(file="resources/gene_names.txt", header=TRUE, sep="\t")
gene_list <- add_rownames(gene_list, "ensemblGeneID")


gene_set <- WT_vs_2D
gene_set <- merge(gene_set, gene_list, by=1)
gene_set <- gene_set %>% filter(!is.na(padj))

#calculate ranks. Sign tells if Fold change is positive or negative

ranks_RNAseq = sign(gene_set$log2FoldChange) * -log10(gene_set$padj)
ranks_RNAseq[ranks_RNAseq ==Inf] <- 350
ranks_RNAseq[ranks_RNAseq ==-Inf] <- -350

#gene names from the TCGA set contain gene name and entrez gene ids separated by ‘|’
# for all subsequent enrichment analysis we need to have just one id.  Separate the names 
# into their two ids.
genenames <- as.character(gene_set$external_gene_name)

names(ranks_RNAseq) <- genenames

fgseaRes <- fgsea(pathways = m_list, 
                  stats = ranks_RNAseq,
                  minSize=3,
                  maxSize=500,
                  nperm=10000)

plotGseaTable(m_list[topPathways], ranks_RNAseq, fgseaRes, gseaParam = 0.5)

```