---
title: "Analysis of DESeq2 results with TFBD motif enrichment"
description: "DEG analysis based on DESeq2 and RcisTarget"
principal investigator: "Joaquín de Navascués"
researcher: "Joaquín de Navascués"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: 'hide'
    theme: readable
    df_print: paged
---
```{r set-publication-theme, echo=FALSE, cache=FALSE}
ggplot2::theme_set(ggpubr::theme_pubr(base_size=10))
```

```{r setup, echo = FALSE, cache = FALSE}
knitr::opts_chunk$set(dev = 'png', 
                      fig.align = 'center', fig.height = 7, fig.width = 8.5, 
                      pdf.options(encoding = "ISOLatin9.enc"),
                      fig.path='integration/figures/', warning=FALSE, message=FALSE)
```

**Libraries needed:**
```{r load_libraries, warning=FALSE, echo=FALSE}
if (!require("librarian")) install.packages("librarian")
# data
librarian::shelf(
  # data manip.
  dplyr, stringr, RcisTarget, data.table, DT,
  # plotting
  ggtheme, ggtext, ggrepel,
# convenience
  here)

if(!exists("gseCP_summarise", mode="function")) source("utils.R")
```

**Set working directory where the script is**
```{r setwd}
if (Sys.getenv("USER")=="JQ") {
  setwd("/Users/JQ/Documents/_CODE REPOS/GitHub/Da_RNAseq")
} else if (Sys.getenv("RSTUDIO")==1) {
  setwd( dirname(rstudioapi::getSourceEditorContext(id = NULL)$path) ) # gets what is in the editor
} else {
  setwd(here::here())
  d <- str_split(getwd(),'/')[[1]][length(str_split(getwd(),'/')[[1]])]
  if (d != 'Da_RNAseq') { stop(
    paste0("Could not set working directory automatically to where this",
           " script resides.\nPlease do `setwd()` manually"))
    }
}
getwd()
```

**Path to definitive images (outside repo):**
```{r define_dir2figs}
figdir <- paste0(c(head(str_split(getwd(),'/')[[1]],-1),
                   paste0(tail(str_split(getwd(),'/')[[1]],1), '_figures')),
                 collapse='/')
dir.create(figdir, showWarnings = FALSE)
```


# Analysis of DESeq2 results with RcisTarget


## 1 Load the data


### 1.1 DEG data


This gets us the DGE data from `DESeq2`, identified by FlyBase/Ensembl ID and gene symbol:
```{r load_DEG_data}
# DEG data
DaKD_deg <- readRDS('output/Control_vs_DaKD.RDS')
DaOE_deg <- readRDS('output/Control_vs_DaOE.RDS')
DaDaOE_deg <- readRDS('output/Control_vs_DaDaOE.RDS')
ScOE_deg <- readRDS('output/Control_vs_ScOE.RDS')
head(
  ScOE_deg %>% dplyr::select(gene_symbol, ensemblGeneID, baseMean, log2FoldChange, padj),
  3
)
```

Now we can define the gene sets as up/down-regulated with our _UAS-x_ transgenes:
```{r}
# gene list by condition and up/downregulated
geneLists <- list(
  daRNAi_dn = make_degset(DaKD_deg, up=FALSE, fc_thresh=1.5)$gene_symbol,
  daRNAi_up = make_degset(DaKD_deg, up=TRUE, fc_thresh=1.5)$gene_symbol,
  daOE_dn = make_degset(DaOE_deg, up=FALSE, fc_thresh=1.5)$gene_symbol,
  daOE_up = make_degset(DaOE_deg, up=TRUE, fc_thresh=1.5)$gene_symbol,
  dadaOE_dn = make_degset(DaDaOE_deg, up=FALSE, fc_thresh=1.5)$gene_symbol,
  dadaOE_up = make_degset(DaDaOE_deg, up=TRUE, fc_thresh=1.5)$gene_symbol,
  scOE_dn = make_degset(ScOE_deg, up=FALSE, fc_thresh=1.5)$gene_symbol,
  scOE_up = make_degset(ScOE_deg, up=TRUE, fc_thresh=1.5)$gene_symbol
)
head(geneLists$daRNAi_dn)
```


### 1.2 Motif databases


Now we need the motif Rankings and Annotations, from the [Stein Aerts' lab website](https://resources.aertslab.org/cistarget/).
Matching them correctly is a bit confusing to me, so I first look carefully at the different options:
I download for _D. melanogaster_ release 6.02:

- VERSION 8 OLD

  - [Motif rankings `.feather` file v8 (old)](https://resources.aertslab.org/cistarget/databases/old/drosophila_melanogaster/dm6/flybase_r6.02/mc8nr/gene_based/dm6-5kb-upstream-full-tx-11species.mc8nr.feather) (1.38Gb)

- VERSION 8 NEW

  - [Motif rankings `.feather` file v8](https://resources.aertslab.org/cistarget/databases/drosophila_melanogaster/dm6/flybase_r6.02/mc8nr/gene_based/dm6-5kb-upstream-full-tx-11species.mc8nr.genes_vs_motifs.rankings.feather) (782Mb) 
  - [Motif annotation table file v8](https://resources.aertslab.org/cistarget/motif2tf/motifs-v8-nr.flybase-m0.001-o0.0.tbl) (40Mb)

- VERSION 10

  - [Motif rankings `.feather` file v10](https://resources.aertslab.org/cistarget/databases/drosophila_melanogaster/dm6/flybase_r6.02/mc_v10_clust/gene_based/dm6_v10_clust.genes_vs_motifs.rankings.feather) (47Mb) 
  - [Motif annotation table file v10](https://resources.aertslab.org/cistarget/motif2tf/motifs-v10nr_clust-nr.flybase-m0.001-o0.0.tbl) (69Mb)

- VERSION 10 'public'

  - [Zipped file with motif annotations for fly, human, mouse and chicken ](https://resources.aertslab.org/cistarget/motif_collections/v10nr_clust_public/v10nr_clust_public.zip) (as individual `.tbl` files; 86Mb)


(**This should be made more 'programmatic' when possible**)

Load the different motif Annotations (mA) and motif Rankings (mR)
```{r}
mA__8new <- importAnnotations("resources/motifdbs/mc8nr/motifs-v8-nr.flybase-m0.001-o0.0.tbl")
mA_10cls <- importAnnotations("resources/motifdbs/mc_v10_clust/motifs-v10nr_clust-nr.flybase-m0.001-o0.0.tbl")
mA_10pub <- importAnnotations("resources/motifdbs/mc_v10_clust/motifs-v10-nr.flybase-m0.00001-o0.0.tbl")

mR__8old <- importRankings("resources/motifdbs/mc8nr/old/dm6-5kb-upstream-full-tx-11species.mc8nr.feather")
mR__8new <- importRankings("resources/motifdbs/mc8nr/dm6-5kb-upstream-full-tx-11species.mc8nr.genes_vs_motifs.rankings.feather")
  mR__8new@rankings <- dplyr::relocate(mR__8new@rankings, motifs)
mR_10cls <- importRankings("resources/motifdbs/mc_v10_clust/dm6_v10_clust.genes_vs_motifs.rankings.feather")
  mR_10cls@rankings <- dplyr::relocate(mR_10cls@rankings, motifs)
```


## 2 Perform motif enrichment with `RcisTarget`


### 2.1 Run with the two versions of the motif annotations


Now I need to see how many motifs are there per database:
```{r}
motif.table <- data.frame(
  motif_annot = c(NA, length(unique(mA__8new$motif)), length(unique(mA_10cls$motif)), length(unique(mA_10pub$motif))),
  motif_ranks = c(nrow(mR__8old), nrow(mR__8new), nrow(mR_10cls), NA)
)
rownames(motif.table) <- c('8old', '8new', '10cls', '10pub')
motif.table
```

So it seems that the combination with more possibilities is to match the v8 motif ranks with the v10 annotation -- the website recommends the opposite, but it is also true that this is in the context of the pySCENE pipeline, which is to build networks with scRNAseq data.

Let's see what we get with a simple comparison:

```{r}
motrich_pure8 <- cisTarget(geneLists, mR__8new, motifAnnot = mA__8new)
motrich_hybr8 <- cisTarget(geneLists, mR__8new, motifAnnot = mA_10cls)
```


### 2.2 Merge the association with TFs


On inspection of the results, most of the information is contained in the 'pure v8' dataframe, whereas the 'hybrid v8-v10' is mostly redundant - but not fully. So I will simply merge them (the connection between motifs and TFs should not depend on how the databases were assembled and the criteria for inclusion is the same).
```{r}
motrich <- motrich_pure8 %>%
  # select only the data I am interested in from the pure v8 enrichment analysis
  dplyr::select(-c(TF_lowConf, rankAtMax, nEnrGenes)) %>%
  # add the data from the hybrid v8-v10 analysis
  tibble::add_column(TF_highConf2 = motrich_hybr8$TF_highConf) %>%
  # remove trailing '. 's and remove source evidence clarifications between brackets
  mutate( across(contains('TF_highConf'), ~ str_remove_all(., " \\(.+\\)|\\. |\\.")) ) %>%
  mutate( across(contains('TF_highConf'), ~ str_remove(., " $")) ) %>%
  # merge v8 / v10 annotations
  unite('TFs', contains('TF_highConf'), sep = "; ") %>%
  # remove piloting/trailing separators
  mutate(TFs = str_remove(TFs, "^; |; $")) %>%
  # split TF names in individual strings
  mutate(TFs = str_split(TFs, '; '))
# remove duplicate TFs
motrich$TFs <- lapply(motrich$TFs, unique)
head(motrich)
```


### 2.3 Associate TFs with TF class using FlyBase's Gene Groups


#### Download the Gene Groups

To assign to gene families, let's start by obtaining the gene family mapping from FlyBase:
```{r, message=FALSE}
geneg.source <- 'http://ftp.flybase.org/releases/FB2023_02/precomputed_files/genes/gene_group_data_fb_2023_02.tsv.gz'
geneg <- data.table::fread(geneg.source)
names(geneg)[[1]] <- 'FB_group_id'
geneg <- geneg %>% dplyr::select( !contains('_id') )
names(geneg) <- c('group_symbol', 'group_name', 'parent_symbol', 'gene_symbol')
nrow(geneg); head(geneg)
```

Now we have a look at how the TFs in `geneg` and `motrich` match each other.

#### Clean up the Gene Group list

Let's start with how many TFs whose motifs were detected by `RcisTarget` do not have a Gene Group:
```{r}
# unique TF names excluding "" for motifs without a _Drosophila_ TF match
detected.tfs <- unique(unlist(motrich$TFs))
detected.tfs <- detected.tfs[ nchar(detected.tfs)>0 ]
# TFs detected in RcisTarget but not present in the group member db:
detected.tfs[!(detected.tfs %in% geneg$gene_symbol)]
```

Of these:

- _h_ is the usual name for _hairy_, which in FB (and therefore in `geneg`) is called _hry_ `-|>` change name in `geneg`
- _CG14440_ and _CG14442_ are homologs of human _ZNF821_, a ZF-C2H2 factor (DIOPT) `-|>` add annotation to `geneg`
- _ocm_ is homolog of human _TBX22_, a T-BOX factor (DIOPT) `-|>` add annotation to `geneg`
- _NK71_ I can't find it, except in this db from the Aerts'lab and poster from the Stark's lab `-|>` remove from `motrich`

```{r}
# add needed gene group annotations
geneg <- geneg %>%
  # change the symbol for _hairy_ from FB official 'hry' to standard 'h'
  mutate(gene_symbol = str_replace(gene_symbol, "^hry$", "h")) %>%
  # add ocm as a T-BOX gene
  add_row(
    geneg %>%
      filter(str_detect(group_name, "T-BOX")) %>%
      summarise(first(.)) %>%
      mutate(gene_symbol='ocm')
    ) %>%
  # add CG14440 and CG14442 as ZF-C2H2 genes
  bind_rows(
    geneg %>%
      filter(str_detect(group_name, "C2H2")) %>%
      head(2) %>%
      mutate(gene_symbol=c('CG14440', 'CG14442'))
    )
# now get rid of all genes that have not been recovered by RcisTarget
geneg <- geneg %>%
  filter(gene_symbol %in% detected.tfs)

# to remove NK71 from the `motrich` TFs, it is better to retrace our steps:
motrich$TFs <- lapply(motrich$TFs, \(x) str_remove(x, 'NK71'))
```

Now we can have a look at genes that have multiple matches in the gene groups:
```{r}
# TFs RcisTarget with multiple gene group memberships:
multifam <- lapply(detected.tfs, function(x) nrow(filter(geneg, gene_symbol==x)))>1
geneg %>%
  filter(gene_symbol %in% detected.tfs[multifam]) %>%
  arrange(gene_symbol)
```

To clean this up, we can simply:

- remove the gene group that does not have a parent symbol
- that will leave us with _Abd-B_, _Myb_, _ham_, _pb_, _sr_, _toy_, _zfh1_

Make gene-specific filters:
```{r}
double_parent_genes <- geneg %>%
  filter(gene_symbol %in% detected.tfs[multifam] & parent_symbol!='') %>%
  filter(geneg %>%
           filter(gene_symbol %in% detected.tfs[multifam] & parent_symbol!='') %>%
           dplyr::select(gene_symbol) %>%
           duplicated()
         ) %>%
  dplyr::select(gene_symbol)
double_parent_genes <- sort(unlist(double_parent_genes))
names(double_parent_genes) <- NULL
# discriminate manually ¯\_(ツ)_/¯ which row to keep:
double_parent_symbol <- c("HOX-C", "KMT", "HTH", "HOX-C", "ULT", "HBTF", "ZN-TF")
```

Apply filters:
```{r}
geneg <- geneg %>%
  # remove the group associations that are multiple and without a parent_symbol
  filter( ! (gene_symbol %in% detected.tfs[multifam] & parent_symbol=='') ) %>%
  # remove the specific associations
  filter( !(gene_symbol==double_parent_genes[[1]] & parent_symbol==double_parent_symbol[[1]]) &
          !(gene_symbol==double_parent_genes[[2]] & parent_symbol==double_parent_symbol[[2]]) &
          !(gene_symbol==double_parent_genes[[3]] & parent_symbol==double_parent_symbol[[3]]) &
          !(gene_symbol==double_parent_genes[[4]] & parent_symbol==double_parent_symbol[[4]]) &
          !(gene_symbol==double_parent_genes[[5]] & parent_symbol==double_parent_symbol[[5]]) &
          !(gene_symbol==double_parent_genes[[6]] & parent_symbol==double_parent_symbol[[6]]) &
          !(gene_symbol==double_parent_genes[[7]] & parent_symbol==double_parent_symbol[[7]]) )
```

#### Associate TF class (Gene Group) to TFs from `RcisTarget`

Now we can provide gene group symbols to the TFs from `RcisTarget`, which is more important than the identity of specific TFs.
```{r}
motrich$TFclass <- lapply(
  motrich$TFs, \(x) unique(unlist(lapply( x, \(y) filter(geneg, gene_symbol==y)$group_symbol)) )
  )
```

As some motifs can be bound by different classes of TFs, we have to do:
```{r}
motrich <- motrich %>% unnest(TFclass)
length(unique(motrich$TFclass))
```

This is still too much for the plot. So I am going to simplify the list of different TF classes by hand:
```{r}
# to map `geneg` descriptions to better, less abundant names:
name2largeclass <- list(
"FORK HEAD BOX TRANSCRIPTION FACTORS" = 'FOX',
"CBF DOMAIN TRANSCRIPTION FACTOR COMPLEX" = 'CBF',
"TATA-BINDING PROTEIN AND TBP-RELATED FACTORS" = 'TBP/TBP-related',
"NK-LIKE HOMEOBOX TRANSCRIPTION FACTORS" = 'Hbox',
"BASIC LEUCINE ZIPPER TRANSCRIPTION FACTORS" = 'bZIP',
"NUCLEAR RECEPTOR SUBFAMILY 0 (LIGAND-INDEPENDENT) TRANSCRIPTION FACTORS" = "NR",
"HIGH MOBILITY GROUP BOX TRANSCRIPTION FACTORS" = "HMG",
"SINE OCULIS HOMEOBOX TRANSCRIPTION FACTORS" = 'Hbox',
"UNCLASSIFIED DNA BINDING DOMAIN TRANSCRIPTION FACTORS" = 'unclassified',
"TEA DOMAIN TRANSCRIPTION FACTORS" = "TEAD",
"MADS-BOX TRANSCRIPTION FACTORS" = "MADS",
"SIRTUIN LYSINE DEACETYLASES" = 'SIRT',
"WNT ENHANCEOSOME" = 'WNTE',
"T-BOX TRANSCRIPTION FACTORS" = 'TBX',
"NUCLEAR RECEPTOR (LIGAND-DEPENDENT) TRANSCRIPTION FACTORS" = "NR",
"SANT-MYB DOMAIN TRANSCRIPTION FACTORS" = "SANT-MYB",
"C2H2 ZINC FINGER TRANSCRIPTION FACTORS" = 'ZnF',
"GATA TRANSCRIPTION FACTORS" = 'GATA',
"BASIC HELIX-LOOP-HELIX TRANSCRIPTION FACTORS" = 'bHLH',
"E2F TRANSCRIPTION FACTORS" = 'E2F',
"GLIAL CELL MISSING TRANSCRIPTION FACTORS" = "GCM",
"PAIRED-LIKE HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox",
"NUCLEAR FACTOR - KAPPA B" = "NFkB",
"POU HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox",
"CUT HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox",
"HOX-LIKE HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox",
"ZINC FINGER HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox",
"TALE HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox",
"NUCLEAR FACTOR OF ACTIVATED T-CELLS TRANSCRIPTION FACTORS" = "NFAT",
"LIM HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox",
"ETS DOMAIN TRANSCRIPTION FACTORS" = "ETS",
"GATOR2 COMPLEX" = "GAT2",
"PAIRED HOMEOBOX TRANSCRIPTION FACTORS" = "Hbox"
)
# to map `geneg` descriptions with symbols
class2name <- (geneg %>%
  dplyr::select(group_symbol, group_name) %>%
  distinct() %>% dplyr::select(group_name))[[1]]
names(class2name) <- (geneg %>%
  dplyr::select(group_symbol, group_name) %>%
  distinct() %>% dplyr::select(group_symbol))[[1]]
# apply to motrich
motrich <- motrich %>%
  mutate(TFLargeClass = recode(TFclass, !!!class2name)) %>%
  mutate(TFLargeClass = recode(TFLargeClass, !!!name2largeclass))
length(unique(motrich$TFLargeClass))
```

Somewhat better... Let's see:


## 3 Plot the results


I want to create a categorical (by experimental condition) dot plot that shows the Y position of the dot as the NES value, and the class of TF as colour. Maybe with jitter, maybe as beeswarm.

The minimal datasets are:
```{r}
# genes up
mots_up <- motrich %>%
  dplyr::select(c(geneSet, NES, TFLargeClass)) %>%
  filter(str_detect(geneSet, 'up'))
mots_up$geneSet <- factor(mots_up$geneSet,
                          levels=c("daRNAi_up", "daOE_up", "dadaOE_up", "scOE_up"),
                          labels=c("*da^RNAi^*", "*da*", "*da:da*", "*scute*"))
# genes down
mots_dn <- motrich %>%
  dplyr::select(c(geneSet, NES, TFLargeClass)) %>%
  filter(str_detect(geneSet, 'dn'))
mots_dn$geneSet <- factor(mots_dn$geneSet,
                          levels=c("daRNAi_dn", "daOE_dn", "dadaOE_dn", "scOE_dn"),
                          labels=c("*da^RNAi^*", "*da*", "*da:da*", "*scute*"))
```

To plot:
```{r}
p <- ggplot(mots_up, aes(x=geneSet, y=NES, fill=TFLargeClass, colour=NULL)) + 
  geom_dotplot(binaxis='y',
               stackdir='center',
               method='histodot',
               position=position_jitter(width=0.1, height=0.1),
               binwidth=0.1,
               dotsize=1.5,
               alpha=0.5,
               stroke=0) +
  theme(axis.text.x = element_markdown())
p
```

```{r}
paste(dplyr::count(mots_up, TFLargeClass))
```

To plot:
```{r}
p <- ggplot(mots_dn, aes(x=geneSet, y=NES, fill=TFLargeClass, colour=NULL)) + 
  geom_dotplot(binaxis='y',
               stackdir='center',
               method='histodot',
               position=position_jitter(width=0.1, height=0.1),
               binwidth=0.1,
               dotsize=1.5,
               alpha=0.5,
               stroke=0) +
  theme(axis.text.x = element_markdown())
p
```

```{r}
paste(dplyr::count(mots_dn, TFLargeClass))
```

TO DO:
make a group of "others" (classes with less than 10)
beautify
create tables with logos
finish and write up.



