---
title: "Analysis of DESeq2 results with TFBD motif enrichment"
description: "DEG analysis based on DESeq2 and RcisTarget"
principal investigator: "Joaquín de Navascués"
researcher: "Joaquín de Navascués"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: 'hide'
    theme: readable
    df_print: paged
---
```{r set-publication-theme, echo=FALSE, cache=FALSE}
ggplot2::theme_set(ggpubr::theme_pubr(base_size=10))
```

```{r setup, echo = FALSE, cache = FALSE}
knitr::opts_chunk$set(dev = 'png', 
                      fig.align = 'center', fig.height = 7, fig.width = 8.5, 
                      pdf.options(encoding = "ISOLatin9.enc"),
                      fig.path='integration/figures/', warning=FALSE, message=FALSE)
```

**Libraries needed:**
```{r load_libraries, warning=FALSE, echo=FALSE}
if (!require("librarian")) install.packages("librarian")
# data
librarian::shelf(
  # data manip.
  dplyr, stringr, RcisTarget, data.table, DT,
  # plotting
  ggtheme, ggtext, ggrepel,
# convenience
  here)

if(!exists("gseCP_summarise", mode="function")) source("utils.R")
```

**Set working directory where the script is**
```{r setwd}
if (Sys.getenv("USER")=="JQ") {
  setwd("/Users/JQ/Documents/_CODE REPOS/GitHub/Da_RNAseq")
} else if (Sys.getenv("RSTUDIO")==1) {
  setwd( dirname(rstudioapi::getSourceEditorContext(id = NULL)$path) ) # gets what is in the editor
} else {
  setwd(here::here())
  d <- str_split(getwd(),'/')[[1]][length(str_split(getwd(),'/')[[1]])]
  if (d != 'Da_RNAseq') { stop(
    paste0("Could not set working directory automatically to where this",
           " script resides.\nPlease do `setwd()` manually"))
    }
}
getwd()
```

**Path to definitive images (outside repo):**
```{r define_dir2figs}
figdir <- paste0(c(head(str_split(getwd(),'/')[[1]],-1),
                   paste0(tail(str_split(getwd(),'/')[[1]],1), '_figures')),
                 collapse='/')
dir.create(figdir, showWarnings = FALSE)
```


# Analysis of DESeq2 results with RcisTarget


## 1 Load the data


This gets us the DGE data from `DESeq2`, identified by FlyBase/Ensembl ID and gene symbol:
```{r load_DEG_data}
# DEG data
DaKD_deg <- readRDS('output/Control_vs_DaKD.RDS')
DaOE_deg <- readRDS('output/Control_vs_DaOE.RDS')
DaDaOE_deg <- readRDS('output/Control_vs_DaDaOE.RDS')
ScOE_deg <- readRDS('output/Control_vs_ScOE.RDS')
head(
  ScOE_deg %>% dplyr::select(gene_symbol, ensemblGeneID, baseMean, log2FoldChange, padj),
  3
)
```

Now we can define the gene sets as up/down-regulated with our _UAS-x_ transgenes:
```{r}
# gene list by condition and up/downregulated
geneLists <- list(
  daRNAi_dn = make_degset(DaKD_deg, up=FALSE, fc_thresh=1.5)$gene_symbol,
  daRNAi_up = make_degset(DaKD_deg, up=TRUE, fc_thresh=1.5)$gene_symbol,
  daOE_dn = make_degset(DaOE_deg, up=FALSE, fc_thresh=1.5)$gene_symbol,
  daOE_up = make_degset(DaOE_deg, up=TRUE, fc_thresh=1.5)$gene_symbol,
  dadaOE_dn = make_degset(DaDaOE_deg, up=FALSE, fc_thresh=1.5)$gene_symbol,
  dadaOE_up = make_degset(DaDaOE_deg, up=TRUE, fc_thresh=1.5)$gene_symbol,
  scOE_dn = make_degset(ScOE_deg, up=FALSE, fc_thresh=1.5)$gene_symbol,
  scOE_up = make_degset(ScOE_deg, up=TRUE, fc_thresh=1.5)$gene_symbol
)
head(geneLists$daRNAi_dn)
```

Now we need the motif Rankings and Annotations, from the [Stein Aerts' lab website](https://resources.aertslab.org/cistarget/).
Matching them correctly is a bit confusing to me, so I first look carefully at the different options:
I download for _D. melanogaster_ release 6.02:

- VERSION 8 OLD

  - [Motif rankings `.feather` file v8 (old)](https://resources.aertslab.org/cistarget/databases/old/drosophila_melanogaster/dm6/flybase_r6.02/mc8nr/gene_based/dm6-5kb-upstream-full-tx-11species.mc8nr.feather) (1.38Gb)

- VERSION 8 NEW

  - [Motif rankings `.feather` file v8](https://resources.aertslab.org/cistarget/databases/drosophila_melanogaster/dm6/flybase_r6.02/mc8nr/gene_based/dm6-5kb-upstream-full-tx-11species.mc8nr.genes_vs_motifs.rankings.feather) (782Mb) 
  - [Motif annotation table file v8](https://resources.aertslab.org/cistarget/motif2tf/motifs-v8-nr.flybase-m0.001-o0.0.tbl) (40Mb)

- VERSION 10

  - [Motif rankings `.feather` file v10](https://resources.aertslab.org/cistarget/databases/drosophila_melanogaster/dm6/flybase_r6.02/mc_v10_clust/gene_based/dm6_v10_clust.genes_vs_motifs.rankings.feather) (47Mb) 
  - [Motif annotation table file v10](https://resources.aertslab.org/cistarget/motif2tf/motifs-v10nr_clust-nr.flybase-m0.001-o0.0.tbl) (69Mb)

- VERSION 10 'public'

  - [Zipped file with motif annotations for fly, human, mouse and chicken ](https://resources.aertslab.org/cistarget/motif_collections/v10nr_clust_public/v10nr_clust_public.zip) (as individual `.tbl` files; 86Mb)

Load the different motif Annotations (mA) and motif Rankings (mR)
```{r}
mA__8new <- importAnnotations("resources/motifdbs/mc8nr/motifs-v8-nr.flybase-m0.001-o0.0.tbl")
mA_10cls <- importAnnotations("resources/motifdbs/mc_v10_clust/motifs-v10nr_clust-nr.flybase-m0.001-o0.0.tbl")
mA_10pub <- importAnnotations("resources/motifdbs/mc_v10_clust/motifs-v10-nr.flybase-m0.00001-o0.0.tbl")

mR__8old <- importRankings("resources/motifdbs/mc8nr/old/dm6-5kb-upstream-full-tx-11species.mc8nr.feather")
mR__8new <- importRankings("resources/motifdbs/mc8nr/dm6-5kb-upstream-full-tx-11species.mc8nr.genes_vs_motifs.rankings.feather")
  mR__8new@rankings <- dplyr::relocate(mR__8new@rankings, motifs)
mR_10cls <- importRankings("resources/motifdbs/mc_v10_clust/dm6_v10_clust.genes_vs_motifs.rankings.feather")
  mR_10cls@rankings <- dplyr::relocate(mR_10cls@rankings, motifs)
```

Now I need to see how many motifs are there per database:
```{r}
motif.table <- data.frame(
  motif_annot = c(NA, length(unique(mA__8new$motif)), length(unique(mA_10cls$motif)), length(unique(mA_10pub$motif))),
  motif_ranks = c(nrow(mR__8old), nrow(mR__8new), nrow(mR_10cls), NA)
)
rownames(motif.table) <- c('8old', '8new', '10cls', '10pub')
motif.table
```

So it seems that the combination with more possibilities is to match the v8 motif ranks with the v10 annotation -- the website recommends the opposite, but it is also true that this is in the context of the pySCENE pipeline, which is to build networks with scRNAseq data.

Let's see what we get with a simple comparison:

```{r}
motrich_pure8 <- cisTarget(geneLists, mR__8new, motifAnnot = mA__8new)
motrich_hybr8 <- cisTarget(geneLists, mR__8new, motifAnnot = mA_10cls)
```


To assign to gene families:
```{r}
geneg <- read.delim('resources/gene_group_data_fb_2023_02.tsv', skip=8)#comment.char = '#')
names(geneg)[[1]] <- 'FB_group_id'
```

