---
title: "Volcano Plots"
output:
  html_document:
    toc: yes
    toc_depth: 2
---

```{r}


library(dplyr)
library(purrr)

setwd('/Users/aleix/Documents/bioinformatics/RNAseq')

WT_vs_2D <- read.table(file="output/differential_expression/2D_vs_WT.txt", header=TRUE, sep="\t")
WT_vs_da <- read.table(file="output/differential_expression/da_vs_WT.txt", header=TRUE, sep="\t")
WT_vs_sc <- read.table(file="output/differential_expression/sc_vs_WT.txt", header=TRUE, sep="\t")
gene_list <- read.table(file="resources/gene_names.txt", header=TRUE, sep="\t")
gene_list <- add_rownames(gene_list, "ensemblGeneID")

tables <- list("2D" = WT_vs_2D, "da" = WT_vs_da, "sc" = WT_vs_sc)
tables <- map(tables, function(df) merge(df, gene_list, by=1))
tables <- map(tables, function(df) mutate(df, mean_tpm = rowMeans(df[,14:16])))

significant_tables <- map(tables, function(df) mutate(df, threshold = padj <0.05 & abs(log2FoldChange) > 1))

selected_genes <- c("ac", "sc", "l(1)sc", "emc", "ase", "da", "phyl", "ttk", "chn", "osa", "esg", "zfh1", "sna", "Sox21a", "fkh", "Dl", "gro", "cbt", "numb", "N", "H", "Su(H)", "neur", "E(spl)m5-HLH", "E(spl)m4-BFM", "E(spl)mbeta-HLH", "E(spl)mdelta-HLH", "E(spl)m2-BFM", "E(spl)m7-HLH", "E(spl)m8-HLH", "E(spl)m3-HLH", "E(spl)malpha-BFM", "E(spl)mgamma-HLH", "E(spl)m6-BFM", "pros", "nub", "Myo31DF", "h", "hdc", "ck", "spen", "Cdk1", "spdo", "polo", "pon", "GATAe", "GATAd", "grn", "pnr", "srp", "ush", "mira", "Zip71B", "Smvt", "Oatp58Dc", "Myc", "bun", "peb", "cic", "Rel", "HmgD", "msn", "klu", "lola", "Lrch", "zip", "insc")

#selected_genes <- c("egr", "grnd", "bsk", "Diap1", "Diap2", "hid", "Dronc", "p53", "Drice", "Duox") # Cell death genes

#Define table to plot
df <- tables$sc

################# VOLCANO PLOT ################# 

'
# Make a basic volcano plot
with(subset(df, padj>=.05 | abs(log2FoldChange)<=1), plot(log2FoldChange, -log10(padj), pch=20, cex=.35, col="#999999", main=list("sc", cex=3.5), xlim=c(min(df$log2FoldChange, na.rm = TRUE),max(df$log2FoldChange, na.rm = TRUE)), ylim=c(0,max(-log10(subset(df, padj > 0)$padj), na.rm = TRUE))))

# Add colored points: red if padj<0.05, orange of log2FC>1, green if both)
with(subset(df, padj<.05 & abs(log2FoldChange)>1), points(log2FoldChange, -log10(padj), pch=20, cex=.35, col="#e2aa3b"))
with(filter(df, external_gene_name %in% selected_genes & padj<.05 & log2FoldChange<(-1)), points(log2FoldChange, -log10(padj), pch=20, cex=.9, col="#721f0f"))
with(filter(df, external_gene_name %in% selected_genes & padj<.05 & log2FoldChange>1), points(log2FoldChange, -log10(padj), pch=20, cex=.9, col="#18631a"))

# Label points with the textxy function from the calibrate plot
library(calibrate)
with(subset(df, external_gene_name %in% selected_genes & padj<.05 & log2FoldChange<(-1)), textxy(log2FoldChange, -log10(padj), labs=external_gene_name, cex=0.9, font=2, col="#721f0f"))
with(subset(df, external_gene_name %in% selected_genes & padj<.05 & log2FoldChange>1), textxy(log2FoldChange, -log10(padj), labs=external_gene_name, cex=0.9, font=2, col="#18631a"))
'
################# SECOND VERSION ################# 

# Make a basic volcano plot
with(subset(df, padj>=.05 | abs(log2FoldChange)<=1), plot(log10(mean_tpm), log2FoldChange, pch=20, cex=.35, col=("#99999965"), main=list("Sc", cex=3.1), ylim=c(min(df$log2FoldChange, na.rm = TRUE),max(df$log2FoldChange, na.rm = TRUE))))

# Add colored points: red if padj<0.05, orange of log2FC>1, green if both)
with(subset(df, padj<.05 & abs(log2FoldChange)>1), points(log10(mean_tpm), log2FoldChange, pch=20, cex=.35, col="#e2aa3b"))
with(filter(df, external_gene_name %in% selected_genes & padj<.05 & log2FoldChange<(-1)), points(log10(mean_tpm), log2FoldChange, pch=20, cex=.9, col="#721f0f"))
with(filter(df, external_gene_name %in% selected_genes & padj<.05 & log2FoldChange>1), points(log10(mean_tpm), log2FoldChange, pch=20, cex=.9, col="#18631a"))
abline(h=c(1,-1), lwd=c(2, 2))

# Label points with the textxy function from the calibrate plot
library(calibrate)
with(subset(df, external_gene_name %in% selected_genes & padj<.05 & log2FoldChange<(-1)), textxy(log10(mean_tpm), log2FoldChange, labs=external_gene_name, cex=0.9, font=2, col="#721f0f"))
with(subset(df, external_gene_name %in% selected_genes & padj<.05 & log2FoldChange>1), textxy(log10(mean_tpm), log2FoldChange, labs=external_gene_name, cex=0.9, font=2, col="#18631a"))

```